{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import re\n",
    "\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.lines import Line2D\n",
    "\n",
    "from collections import Counter\n",
    "\n",
    "all_markers = list(Line2D.markers.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyzing NOSC of the E. coli proteome\n",
    "This notebook calculates the condition dependent proteome NOSC for the datasets aggregated by Chure & Bellivue et al. Cell Systems 2021. To calculate the proteome NOSC we need near-complete proteome datasets such that the uncertainty related to unmeasured proteins is small. Recent datasets of this quality were collected by Griffin Chure and Nathan Bellivue. \n",
    "\n",
    "Since proteome NOSC is a weighted average of contributions from individual proteins, we do not need absolute measurements with real units. Rather, we need accurate and complete compositional data, i.e. protein A makes P% of the proteome. \n",
    "\n",
    "# Differences between lab practices\n",
    "Valgepea 2013 uses data collected in Valgepea 2010, where E. coli was cultured in a custom minimal medium documented in Nahku et al. 2010. Peebo et al. 2015 is from the same group (Vilu) and appears to use the same growth medium base with some amino acid supplementation. \n",
    "\n",
    "Minimal media used by Schmidt et al. 2015 uses an M9 base that is distinct from the medium in the Vilu group. \n",
    "\n",
    "# Known issues\n",
    "I am using MG1655 coding sequences for all samples, but Schmidt et al. and Peebo et al. are working with BW25113. The BW25113 is derived from MG1655 with a small number deletions and other changes documented in the genome announcement. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4200154/\n",
    "\n",
    "# Some additional E. coli data sources\n",
    "* Hui et al. MSB 2015 is from Hwa & Williamson groups\n",
    "* Maser et al. 2021 from Vilu and Nahku groups\n",
    "* Mori et al. 2022 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_abund_df = pd.read_csv('../data/proteomes/Coli/Chure_compiled_absolute_measurements.csv', index_col=0).reset_index().set_index('b_number')\n",
    "nosc_df = pd.read_csv('../data/genomes/Coli/MG1655/MG1655_ref_prot_NOSC.csv').set_index('b_number')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aa_seq</th>\n",
       "      <th>num_aas</th>\n",
       "      <th>mw_daltons</th>\n",
       "      <th>transmembrane_aas</th>\n",
       "      <th>fraction_transmembrane</th>\n",
       "      <th>primary_accession</th>\n",
       "      <th>accessions</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>description</th>\n",
       "      <th>locus_tags</th>\n",
       "      <th>...</th>\n",
       "      <th>COG_IDs</th>\n",
       "      <th>KEGG_IDs</th>\n",
       "      <th>isoform_accessions</th>\n",
       "      <th>KEGG_path_IDs</th>\n",
       "      <th>KEGG_pathways</th>\n",
       "      <th>Ce_total</th>\n",
       "      <th>NC</th>\n",
       "      <th>NOSC</th>\n",
       "      <th>eC_ratio</th>\n",
       "      <th>organism</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b4586</th>\n",
       "      <td>MRLHVKLKEFLSMFFMAILFFPAFNASLFFTGVKPLYSIIKCSTEI...</td>\n",
       "      <td>159</td>\n",
       "      <td>18118</td>\n",
       "      <td>60</td>\n",
       "      <td>0.377358</td>\n",
       "      <td>A5A605</td>\n",
       "      <td>A5A605,P71282</td>\n",
       "      <td>ykfM</td>\n",
       "      <td>Uncharacterized protein YkfM</td>\n",
       "      <td>ykfM,b4586</td>\n",
       "      <td>...</td>\n",
       "      <td>ENOG5031M8I</td>\n",
       "      <td>eco:b4586</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-301.25</td>\n",
       "      <td>862.0</td>\n",
       "      <td>-0.349478</td>\n",
       "      <td>4.349478</td>\n",
       "      <td>coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b1252</th>\n",
       "      <td>MTLDLPRRFPWPTLLSVCIHGAVVAGLLYTSVHQVIELPAPAQPIS...</td>\n",
       "      <td>239</td>\n",
       "      <td>26094</td>\n",
       "      <td>31</td>\n",
       "      <td>0.129707</td>\n",
       "      <td>P02929</td>\n",
       "      <td>P02929,P76831,P94719,P94722,P94726,P94728,P947...</td>\n",
       "      <td>tonB</td>\n",
       "      <td>Protein TonB</td>\n",
       "      <td>tonB,exbA,b1252,JW5195</td>\n",
       "      <td>...</td>\n",
       "      <td>COG0810</td>\n",
       "      <td>ecj:JW5195,eco:b1252</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-244.16</td>\n",
       "      <td>1173.0</td>\n",
       "      <td>-0.208150</td>\n",
       "      <td>4.208150</td>\n",
       "      <td>coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3635</th>\n",
       "      <td>MPELPEVETSRRGIEPHLVGATILHAVVRNGRLRWPVSEEIYRLSD...</td>\n",
       "      <td>269</td>\n",
       "      <td>30290</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>P05523</td>\n",
       "      <td>P05523,Q2M7U9</td>\n",
       "      <td>mutM</td>\n",
       "      <td>Formamidopyrimidine-DNA glycosylase</td>\n",
       "      <td>mutM,fpg,b3635,JW3610</td>\n",
       "      <td>...</td>\n",
       "      <td>COG0266</td>\n",
       "      <td>ecj:JW3610,eco:b3635</td>\n",
       "      <td>NaN</td>\n",
       "      <td>path:eco03410</td>\n",
       "      <td>Base excision repair - Escherichia coli K-12 M...</td>\n",
       "      <td>-231.79</td>\n",
       "      <td>1358.0</td>\n",
       "      <td>-0.170685</td>\n",
       "      <td>4.170685</td>\n",
       "      <td>coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b1709</th>\n",
       "      <td>MSIVMQLQDVAESTRLGPLSGEVRAGEILHLVGPNGAGKSTLLARM...</td>\n",
       "      <td>249</td>\n",
       "      <td>27081</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>P06611</td>\n",
       "      <td>P06611</td>\n",
       "      <td>btuD</td>\n",
       "      <td>Vitamin B12 import ATP-binding protein BtuD</td>\n",
       "      <td>btuD,b1709,JW1699</td>\n",
       "      <td>...</td>\n",
       "      <td>COG4138</td>\n",
       "      <td>ecj:JW1699,eco:b1709</td>\n",
       "      <td>NaN</td>\n",
       "      <td>path:eco02010</td>\n",
       "      <td>ABC transporters - Escherichia coli K-12 MG1655</td>\n",
       "      <td>-157.82</td>\n",
       "      <td>1189.0</td>\n",
       "      <td>-0.132733</td>\n",
       "      <td>4.132733</td>\n",
       "      <td>coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2306</th>\n",
       "      <td>MSENKLNVIDLHKRYGEHEVLKGVSLQANAGDVISIIGSSGSGKST...</td>\n",
       "      <td>257</td>\n",
       "      <td>28653</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>P07109</td>\n",
       "      <td>P07109,P77299</td>\n",
       "      <td>hisP</td>\n",
       "      <td>Histidine transport ATP-binding protein HisP</td>\n",
       "      <td>hisP,b2306,JW2303</td>\n",
       "      <td>...</td>\n",
       "      <td>COG4598</td>\n",
       "      <td>ecj:JW2303,eco:b2306</td>\n",
       "      <td>NaN</td>\n",
       "      <td>path:eco02010</td>\n",
       "      <td>ABC transporters - Escherichia coli K-12 MG1655</td>\n",
       "      <td>-201.82</td>\n",
       "      <td>1266.0</td>\n",
       "      <td>-0.159415</td>\n",
       "      <td>4.159415</td>\n",
       "      <td>coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4346-P15005-2</th>\n",
       "      <td>MSKTESYCLEDALNDLFIPETTIETILKRLTIKKNIILQGPPGVGK...</td>\n",
       "      <td>298</td>\n",
       "      <td>53157</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>P15005-2</td>\n",
       "      <td>P15005,Q2M5X0</td>\n",
       "      <td>mcrB</td>\n",
       "      <td>Type IV methyl-directed restriction enzyme Eco...</td>\n",
       "      <td>mcrB,rglB,b4346,JW5871</td>\n",
       "      <td>...</td>\n",
       "      <td>COG1401</td>\n",
       "      <td>ecj:JW5871,eco:b4346</td>\n",
       "      <td>P15005-1,P15005-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-239.32</td>\n",
       "      <td>1563.0</td>\n",
       "      <td>-0.153116</td>\n",
       "      <td>4.153116</td>\n",
       "      <td>coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b1120-P75960-2</th>\n",
       "      <td>MEKPRVLVLTGAGISAESGIRTFRAADGLWEEHRVEDVATPEGFDR...</td>\n",
       "      <td>242</td>\n",
       "      <td>31464</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>P75960-2</td>\n",
       "      <td>P75960</td>\n",
       "      <td>cobB</td>\n",
       "      <td>NAD-dependent protein deacylase</td>\n",
       "      <td>cobB,ycfY,b1120,JW1106</td>\n",
       "      <td>...</td>\n",
       "      <td>COG0846</td>\n",
       "      <td>ecj:JW1106,eco:b1120</td>\n",
       "      <td>P75960-1,P75960-2</td>\n",
       "      <td>path:eco00760,path:eco01100</td>\n",
       "      <td>Nicotinate and nicotinamide metabolism - Esche...</td>\n",
       "      <td>-133.51</td>\n",
       "      <td>1189.0</td>\n",
       "      <td>-0.112288</td>\n",
       "      <td>4.112288</td>\n",
       "      <td>coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0149-P02919-2</th>\n",
       "      <td>MPRKGKGKGKGRKPRGKRGWLWLLLKLAIVFAVLIAIYGVYLDQKI...</td>\n",
       "      <td>799</td>\n",
       "      <td>94293</td>\n",
       "      <td>23</td>\n",
       "      <td>0.027251</td>\n",
       "      <td>P02919-2</td>\n",
       "      <td>P02919,P75664</td>\n",
       "      <td>mrcB</td>\n",
       "      <td>Penicillin-binding protein 1B</td>\n",
       "      <td>mrcB,pbpF,ponB,b0149,JW0145</td>\n",
       "      <td>...</td>\n",
       "      <td>COG0744</td>\n",
       "      <td>ecj:JW0145,eco:b0149</td>\n",
       "      <td>P02919-1,P02919-2</td>\n",
       "      <td>path:eco00550,path:eco01100</td>\n",
       "      <td>Peptidoglycan biosynthesis - Escherichia coli ...</td>\n",
       "      <td>-559.41</td>\n",
       "      <td>3952.0</td>\n",
       "      <td>-0.141551</td>\n",
       "      <td>4.141551</td>\n",
       "      <td>coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2234-P00452-2</th>\n",
       "      <td>MDWAAEGLHNVSISQVELRSHIQFYDGIKTSDIHETIIKAAADLIS...</td>\n",
       "      <td>736</td>\n",
       "      <td>85775</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>P00452-2</td>\n",
       "      <td>P00452,P78088,P78177</td>\n",
       "      <td>nrdA</td>\n",
       "      <td>Ribonucleoside-diphosphate reductase 1 subunit...</td>\n",
       "      <td>nrdA,dnaF,b2234,JW2228</td>\n",
       "      <td>...</td>\n",
       "      <td>COG0209</td>\n",
       "      <td>ecj:JW2228,eco:b2234</td>\n",
       "      <td>P00452-1,P00452-2</td>\n",
       "      <td>path:eco00230,path:eco00240,path:eco01100,path...</td>\n",
       "      <td>Purine metabolism - Escherichia coli K-12 MG16...</td>\n",
       "      <td>-478.53</td>\n",
       "      <td>3700.0</td>\n",
       "      <td>-0.129332</td>\n",
       "      <td>4.129332</td>\n",
       "      <td>coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4795-P0DSH3-2</th>\n",
       "      <td>MNIFKPISYIASLAPREVTLLALV</td>\n",
       "      <td>24</td>\n",
       "      <td>8757</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>P0DSH3-2</td>\n",
       "      <td>P0DSH3</td>\n",
       "      <td>yibX</td>\n",
       "      <td>Protein YibX</td>\n",
       "      <td>yibX,b4795</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P0DSH3-1,P0DSH3-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-53.96</td>\n",
       "      <td>125.0</td>\n",
       "      <td>-0.431680</td>\n",
       "      <td>4.431680</td>\n",
       "      <td>coli</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4412 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           aa_seq  num_aas  \\\n",
       "b_number                                                                     \n",
       "b4586           MRLHVKLKEFLSMFFMAILFFPAFNASLFFTGVKPLYSIIKCSTEI...      159   \n",
       "b1252           MTLDLPRRFPWPTLLSVCIHGAVVAGLLYTSVHQVIELPAPAQPIS...      239   \n",
       "b3635           MPELPEVETSRRGIEPHLVGATILHAVVRNGRLRWPVSEEIYRLSD...      269   \n",
       "b1709           MSIVMQLQDVAESTRLGPLSGEVRAGEILHLVGPNGAGKSTLLARM...      249   \n",
       "b2306           MSENKLNVIDLHKRYGEHEVLKGVSLQANAGDVISIIGSSGSGKST...      257   \n",
       "...                                                           ...      ...   \n",
       "b4346-P15005-2  MSKTESYCLEDALNDLFIPETTIETILKRLTIKKNIILQGPPGVGK...      298   \n",
       "b1120-P75960-2  MEKPRVLVLTGAGISAESGIRTFRAADGLWEEHRVEDVATPEGFDR...      242   \n",
       "b0149-P02919-2  MPRKGKGKGKGRKPRGKRGWLWLLLKLAIVFAVLIAIYGVYLDQKI...      799   \n",
       "b2234-P00452-2  MDWAAEGLHNVSISQVELRSHIQFYDGIKTSDIHETIIKAAADLIS...      736   \n",
       "b4795-P0DSH3-2                           MNIFKPISYIASLAPREVTLLALV       24   \n",
       "\n",
       "                mw_daltons  transmembrane_aas  fraction_transmembrane  \\\n",
       "b_number                                                                \n",
       "b4586                18118                 60                0.377358   \n",
       "b1252                26094                 31                0.129707   \n",
       "b3635                30290                  0                0.000000   \n",
       "b1709                27081                  0                0.000000   \n",
       "b2306                28653                  0                0.000000   \n",
       "...                    ...                ...                     ...   \n",
       "b4346-P15005-2       53157                  0                0.000000   \n",
       "b1120-P75960-2       31464                  0                0.000000   \n",
       "b0149-P02919-2       94293                 23                0.027251   \n",
       "b2234-P00452-2       85775                  0                0.000000   \n",
       "b4795-P0DSH3-2        8757                  0                0.000000   \n",
       "\n",
       "               primary_accession  \\\n",
       "b_number                           \n",
       "b4586                     A5A605   \n",
       "b1252                     P02929   \n",
       "b3635                     P05523   \n",
       "b1709                     P06611   \n",
       "b2306                     P07109   \n",
       "...                          ...   \n",
       "b4346-P15005-2          P15005-2   \n",
       "b1120-P75960-2          P75960-2   \n",
       "b0149-P02919-2          P02919-2   \n",
       "b2234-P00452-2          P00452-2   \n",
       "b4795-P0DSH3-2          P0DSH3-2   \n",
       "\n",
       "                                                       accessions gene_name  \\\n",
       "b_number                                                                      \n",
       "b4586                                               A5A605,P71282      ykfM   \n",
       "b1252           P02929,P76831,P94719,P94722,P94726,P94728,P947...      tonB   \n",
       "b3635                                               P05523,Q2M7U9      mutM   \n",
       "b1709                                                      P06611      btuD   \n",
       "b2306                                               P07109,P77299      hisP   \n",
       "...                                                           ...       ...   \n",
       "b4346-P15005-2                                      P15005,Q2M5X0      mcrB   \n",
       "b1120-P75960-2                                             P75960      cobB   \n",
       "b0149-P02919-2                                      P02919,P75664      mrcB   \n",
       "b2234-P00452-2                               P00452,P78088,P78177      nrdA   \n",
       "b4795-P0DSH3-2                                             P0DSH3      yibX   \n",
       "\n",
       "                                                      description  \\\n",
       "b_number                                                            \n",
       "b4586                                Uncharacterized protein YkfM   \n",
       "b1252                                                Protein TonB   \n",
       "b3635                         Formamidopyrimidine-DNA glycosylase   \n",
       "b1709                 Vitamin B12 import ATP-binding protein BtuD   \n",
       "b2306                Histidine transport ATP-binding protein HisP   \n",
       "...                                                           ...   \n",
       "b4346-P15005-2  Type IV methyl-directed restriction enzyme Eco...   \n",
       "b1120-P75960-2                    NAD-dependent protein deacylase   \n",
       "b0149-P02919-2                      Penicillin-binding protein 1B   \n",
       "b2234-P00452-2  Ribonucleoside-diphosphate reductase 1 subunit...   \n",
       "b4795-P0DSH3-2                                       Protein YibX   \n",
       "\n",
       "                                 locus_tags  ...      COG_IDs  \\\n",
       "b_number                                     ...                \n",
       "b4586                            ykfM,b4586  ...  ENOG5031M8I   \n",
       "b1252                tonB,exbA,b1252,JW5195  ...      COG0810   \n",
       "b3635                 mutM,fpg,b3635,JW3610  ...      COG0266   \n",
       "b1709                     btuD,b1709,JW1699  ...      COG4138   \n",
       "b2306                     hisP,b2306,JW2303  ...      COG4598   \n",
       "...                                     ...  ...          ...   \n",
       "b4346-P15005-2       mcrB,rglB,b4346,JW5871  ...      COG1401   \n",
       "b1120-P75960-2       cobB,ycfY,b1120,JW1106  ...      COG0846   \n",
       "b0149-P02919-2  mrcB,pbpF,ponB,b0149,JW0145  ...      COG0744   \n",
       "b2234-P00452-2       nrdA,dnaF,b2234,JW2228  ...      COG0209   \n",
       "b4795-P0DSH3-2                   yibX,b4795  ...          NaN   \n",
       "\n",
       "                            KEGG_IDs isoform_accessions  \\\n",
       "b_number                                                  \n",
       "b4586                      eco:b4586                NaN   \n",
       "b1252           ecj:JW5195,eco:b1252                NaN   \n",
       "b3635           ecj:JW3610,eco:b3635                NaN   \n",
       "b1709           ecj:JW1699,eco:b1709                NaN   \n",
       "b2306           ecj:JW2303,eco:b2306                NaN   \n",
       "...                              ...                ...   \n",
       "b4346-P15005-2  ecj:JW5871,eco:b4346  P15005-1,P15005-2   \n",
       "b1120-P75960-2  ecj:JW1106,eco:b1120  P75960-1,P75960-2   \n",
       "b0149-P02919-2  ecj:JW0145,eco:b0149  P02919-1,P02919-2   \n",
       "b2234-P00452-2  ecj:JW2228,eco:b2234  P00452-1,P00452-2   \n",
       "b4795-P0DSH3-2                   NaN  P0DSH3-1,P0DSH3-2   \n",
       "\n",
       "                                                    KEGG_path_IDs  \\\n",
       "b_number                                                            \n",
       "b4586                                                         NaN   \n",
       "b1252                                                         NaN   \n",
       "b3635                                               path:eco03410   \n",
       "b1709                                               path:eco02010   \n",
       "b2306                                               path:eco02010   \n",
       "...                                                           ...   \n",
       "b4346-P15005-2                                                NaN   \n",
       "b1120-P75960-2                        path:eco00760,path:eco01100   \n",
       "b0149-P02919-2                        path:eco00550,path:eco01100   \n",
       "b2234-P00452-2  path:eco00230,path:eco00240,path:eco01100,path...   \n",
       "b4795-P0DSH3-2                                                NaN   \n",
       "\n",
       "                                                    KEGG_pathways Ce_total  \\\n",
       "b_number                                                                     \n",
       "b4586                                                         NaN  -301.25   \n",
       "b1252                                                         NaN  -244.16   \n",
       "b3635           Base excision repair - Escherichia coli K-12 M...  -231.79   \n",
       "b1709             ABC transporters - Escherichia coli K-12 MG1655  -157.82   \n",
       "b2306             ABC transporters - Escherichia coli K-12 MG1655  -201.82   \n",
       "...                                                           ...      ...   \n",
       "b4346-P15005-2                                                NaN  -239.32   \n",
       "b1120-P75960-2  Nicotinate and nicotinamide metabolism - Esche...  -133.51   \n",
       "b0149-P02919-2  Peptidoglycan biosynthesis - Escherichia coli ...  -559.41   \n",
       "b2234-P00452-2  Purine metabolism - Escherichia coli K-12 MG16...  -478.53   \n",
       "b4795-P0DSH3-2                                                NaN   -53.96   \n",
       "\n",
       "                    NC      NOSC  eC_ratio  organism  \n",
       "b_number                                              \n",
       "b4586            862.0 -0.349478  4.349478      coli  \n",
       "b1252           1173.0 -0.208150  4.208150      coli  \n",
       "b3635           1358.0 -0.170685  4.170685      coli  \n",
       "b1709           1189.0 -0.132733  4.132733      coli  \n",
       "b2306           1266.0 -0.159415  4.159415      coli  \n",
       "...                ...       ...       ...       ...  \n",
       "b4346-P15005-2  1563.0 -0.153116  4.153116      coli  \n",
       "b1120-P75960-2  1189.0 -0.112288  4.112288      coli  \n",
       "b0149-P02919-2  3952.0 -0.141551  4.141551      coli  \n",
       "b2234-P00452-2  3700.0 -0.129332  4.129332      coli  \n",
       "b4795-P0DSH3-2   125.0 -0.431680  4.431680      coli  \n",
       "\n",
       "[4412 rows x 21 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nosc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>strain</th>\n",
       "      <th>condition</th>\n",
       "      <th>growth_rate_hr</th>\n",
       "      <th>copies_per_cell</th>\n",
       "      <th>fg_per_cell</th>\n",
       "      <th>NOSC</th>\n",
       "      <th>NC_per</th>\n",
       "      <th>num_aas</th>\n",
       "      <th>mw_daltons</th>\n",
       "      <th>organism_key</th>\n",
       "      <th>species</th>\n",
       "      <th>majority_protein_ids</th>\n",
       "      <th>fraction_transmembrane</th>\n",
       "      <th>Ce_per</th>\n",
       "      <th>Ce_total</th>\n",
       "      <th>NC_total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0167</th>\n",
       "      <td>li_2014</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>MOPS complete</td>\n",
       "      <td>1.934364</td>\n",
       "      <td>206.892387</td>\n",
       "      <td>0.035165</td>\n",
       "      <td>-0.126187</td>\n",
       "      <td>4569.0</td>\n",
       "      <td>890</td>\n",
       "      <td>102390</td>\n",
       "      <td>coli</td>\n",
       "      <td>E. coli</td>\n",
       "      <td>P27249</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-576.55</td>\n",
       "      <td>-119283.805932</td>\n",
       "      <td>945291.317842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0167</th>\n",
       "      <td>li_2014</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>MOPS complete without methionine</td>\n",
       "      <td>1.569390</td>\n",
       "      <td>122.291713</td>\n",
       "      <td>0.020786</td>\n",
       "      <td>-0.126187</td>\n",
       "      <td>4569.0</td>\n",
       "      <td>890</td>\n",
       "      <td>102390</td>\n",
       "      <td>coli</td>\n",
       "      <td>E. coli</td>\n",
       "      <td>P27249</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-576.55</td>\n",
       "      <td>-70507.286956</td>\n",
       "      <td>558750.835314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0167</th>\n",
       "      <td>li_2014</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>MOPS minimal</td>\n",
       "      <td>0.738700</td>\n",
       "      <td>190.617683</td>\n",
       "      <td>0.032399</td>\n",
       "      <td>-0.126187</td>\n",
       "      <td>4569.0</td>\n",
       "      <td>890</td>\n",
       "      <td>102390</td>\n",
       "      <td>coli</td>\n",
       "      <td>E. coli</td>\n",
       "      <td>P27249</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-576.55</td>\n",
       "      <td>-109900.625254</td>\n",
       "      <td>870932.194582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0167</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>glucose_minimal</td>\n",
       "      <td>0.210000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.126187</td>\n",
       "      <td>4569.0</td>\n",
       "      <td>890</td>\n",
       "      <td>102390</td>\n",
       "      <td>coli</td>\n",
       "      <td>E. coli</td>\n",
       "      <td>P27249</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-576.55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0167</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>glucose_minimal</td>\n",
       "      <td>0.220000</td>\n",
       "      <td>4.488067</td>\n",
       "      <td>0.000763</td>\n",
       "      <td>-0.126187</td>\n",
       "      <td>4569.0</td>\n",
       "      <td>890</td>\n",
       "      <td>102390</td>\n",
       "      <td>coli</td>\n",
       "      <td>E. coli</td>\n",
       "      <td>P27249</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-576.55</td>\n",
       "      <td>-2587.595263</td>\n",
       "      <td>20505.979980</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             dataset   strain                         condition  \\\n",
       "b_number                                                          \n",
       "b0167        li_2014   MG1655                     MOPS complete   \n",
       "b0167        li_2014   MG1655  MOPS complete without methionine   \n",
       "b0167        li_2014   MG1655                      MOPS minimal   \n",
       "b0167     peebo_2015  BW25113                   glucose_minimal   \n",
       "b0167     peebo_2015  BW25113                   glucose_minimal   \n",
       "\n",
       "          growth_rate_hr  copies_per_cell  fg_per_cell      NOSC  NC_per  \\\n",
       "b_number                                                                   \n",
       "b0167           1.934364       206.892387     0.035165 -0.126187  4569.0   \n",
       "b0167           1.569390       122.291713     0.020786 -0.126187  4569.0   \n",
       "b0167           0.738700       190.617683     0.032399 -0.126187  4569.0   \n",
       "b0167           0.210000              NaN          NaN -0.126187  4569.0   \n",
       "b0167           0.220000         4.488067     0.000763 -0.126187  4569.0   \n",
       "\n",
       "          num_aas  mw_daltons organism_key  species majority_protein_ids  \\\n",
       "b_number                                                                   \n",
       "b0167         890      102390         coli  E. coli               P27249   \n",
       "b0167         890      102390         coli  E. coli               P27249   \n",
       "b0167         890      102390         coli  E. coli               P27249   \n",
       "b0167         890      102390         coli  E. coli               P27249   \n",
       "b0167         890      102390         coli  E. coli               P27249   \n",
       "\n",
       "          fraction_transmembrane  Ce_per       Ce_total       NC_total  \n",
       "b_number                                                                \n",
       "b0167                        0.0 -576.55 -119283.805932  945291.317842  \n",
       "b0167                        0.0 -576.55  -70507.286956  558750.835314  \n",
       "b0167                        0.0 -576.55 -109900.625254  870932.194582  \n",
       "b0167                        0.0 -576.55            NaN            NaN  \n",
       "b0167                        0.0 -576.55   -2587.595263   20505.979980  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Take the mean of replicates for the same gene. \n",
    "# Note 1: have to do this because didn't report which measurement is from which replicate.\n",
    "# Note 2: the Schmidt data is already aggregated across two measurement methods.\n",
    "# NB: Assuming same dataset, strain, cond, gene and growth rate implies replicate. \n",
    "counts = raw_abund_df.reset_index().groupby('dataset,strain,condition,b_number,growth_rate_hr'.split(',')).mean()\n",
    "# renaming tot_per_cell for clarity and to match how we treat other datasets\n",
    "abund_df = counts.reset_index().set_index('b_number').rename(\n",
    "    columns=dict(tot_per_cell='copies_per_cell'))\n",
    "\n",
    "# Keep only the b-numbers that we have abundance and NOSC data for.\n",
    "overlapping_idx = set(nosc_df.index.values).intersection(abund_df.index.values)\n",
    "abund_df = abund_df.loc[overlapping_idx].copy()\n",
    "nosc_df = nosc_df.loc[overlapping_idx].copy()\n",
    "\n",
    "# Add NOSC and NC data to the abundance dataframe\n",
    "abund_df['NOSC'] = nosc_df.loc[abund_df.index.values].NOSC\n",
    "abund_df['NC_per'] = nosc_df.loc[abund_df.index.values].NC\n",
    "abund_df['num_aas'] = nosc_df.loc[abund_df.index.values].num_aas\n",
    "abund_df['mw_daltons'] = nosc_df.loc[abund_df.index.values].mw_daltons\n",
    "abund_df['organism_key'] = 'coli'\n",
    "abund_df['species'] = 'E. coli'\n",
    "abund_df['majority_protein_ids'] = nosc_df.loc[abund_df.index.values].primary_accession\n",
    "abund_df['fraction_transmembrane'] = nosc_df.loc[abund_df.index.values].fraction_transmembrane\n",
    "\n",
    "# Formal Carbon-bound e- per protein copy\n",
    "abund_df['Ce_per'] = nosc_df.loc[abund_df.index].Ce_total\n",
    "# Total Carbon-bound e-/cell on this protein, i.e. weighted by copies/cell \n",
    "abund_df['Ce_total'] = abund_df.Ce_per * abund_df.copies_per_cell\n",
    "# Total Carbon atoms on this protein, i.e. weighted by copies/cell \n",
    "abund_df['NC_total'] = abund_df.NC_per * abund_df.copies_per_cell\n",
    "abund_df\n",
    "abund_df.to_csv('../data/proteomes/Coli/Chure_mean_absolute_measurements.csv')\n",
    "abund_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>M</th>\n",
       "      <th>N</th>\n",
       "      <th>T</th>\n",
       "      <th>L</th>\n",
       "      <th>P</th>\n",
       "      <th>E</th>\n",
       "      <th>Q</th>\n",
       "      <th>Y</th>\n",
       "      <th>A</th>\n",
       "      <th>G</th>\n",
       "      <th>...</th>\n",
       "      <th>V</th>\n",
       "      <th>W</th>\n",
       "      <th>R</th>\n",
       "      <th>D</th>\n",
       "      <th>I</th>\n",
       "      <th>K</th>\n",
       "      <th>H</th>\n",
       "      <th>F</th>\n",
       "      <th>S</th>\n",
       "      <th>U</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0167</th>\n",
       "      <td>19</td>\n",
       "      <td>28.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>122.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>...</td>\n",
       "      <td>50.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0284</th>\n",
       "      <td>19</td>\n",
       "      <td>26.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>...</td>\n",
       "      <td>51.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3725</th>\n",
       "      <td>8</td>\n",
       "      <td>12.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b1760</th>\n",
       "      <td>1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2597</th>\n",
       "      <td>4</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           M     N     T      L     P     E     Q     Y     A     G  ...  \\\n",
       "b_number                                                             ...   \n",
       "b0167     19  28.0  48.0  122.0  40.0  62.0  54.0  16.0  65.0  46.0  ...   \n",
       "b0284     19  26.0  52.0   60.0  38.0  41.0  27.0  17.0  95.0  71.0  ...   \n",
       "b3725      8  12.0  17.0   27.0  10.0  16.0  14.0   9.0  19.0  16.0  ...   \n",
       "b1760      1   4.0   3.0    8.0   8.0   4.0   7.0   3.0   8.0   4.0  ...   \n",
       "b2597      4   8.0   8.0    9.0   4.0  12.0   6.0   1.0  11.0   4.0  ...   \n",
       "\n",
       "             V     W     R     D     I     K     H     F     S    U  \n",
       "b_number                                                             \n",
       "b0167     50.0  16.0  77.0  52.0  46.0  22.0  38.0  38.0  42.0  0.0  \n",
       "b0284     51.0   8.0  39.0  43.0  34.0  32.0  19.0  18.0  36.0  0.0  \n",
       "b3725     11.0   2.0  13.0  15.0  18.0  16.0   4.0  12.0  14.0  0.0  \n",
       "b1760     10.0   1.0   5.0   4.0   3.0   3.0   1.0   3.0   6.0  0.0  \n",
       "b2597      7.0   1.0   5.0   4.0   9.0   9.0   5.0   2.0   4.0  0.0  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# amino acid counts per protein\n",
    "aa_counts = [Counter(a) for a in nosc_df.aa_seq]\n",
    "aa_counts_df = pd.DataFrame(aa_counts, index=nosc_df.index).replace({np.NaN: 0})\n",
    "aa_counts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the fast growth sample from schmidt\n",
    "mask = np.logical_and(abund_df.dataset == 'schmidt_2016',\n",
    "                      abund_df.condition == 'chemostat_u0.5')\n",
    "mean_abund_fast_growth = abund_df[mask].reset_index().set_index('b_number').replace({np.NaN: 0})\n",
    "\n",
    "overlapping_ids = set(mean_abund_fast_growth.index).intersection(aa_counts_df.index)\n",
    "tmp = aa_counts_df.loc[overlapping_ids].multiply(mean_abund_fast_growth.loc[overlapping_ids].copies_per_cell, axis=0).sum()\n",
    "expression_weighted_aas = tmp / tmp.sum()\n",
    "expression_weighted_aas.name = 'aa_freq'\n",
    "\n",
    "expression_weighted_aas.to_csv('../data/proteomes/Coli/Schmidt_mu0.5_expression_weighted_aa_freqs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>strain</th>\n",
       "      <th>condition</th>\n",
       "      <th>growth_rate_hr</th>\n",
       "      <th>Ce_total</th>\n",
       "      <th>NC_total</th>\n",
       "      <th>proteome_NOSC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>li_2014</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>MOPS complete</td>\n",
       "      <td>1.934364</td>\n",
       "      <td>-2.873416e+09</td>\n",
       "      <td>1.979388e+10</td>\n",
       "      <td>-0.145167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>li_2014</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>MOPS complete without methionine</td>\n",
       "      <td>1.569390</td>\n",
       "      <td>-2.085635e+09</td>\n",
       "      <td>1.456026e+10</td>\n",
       "      <td>-0.143242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>li_2014</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>MOPS minimal</td>\n",
       "      <td>0.738700</td>\n",
       "      <td>-1.020722e+09</td>\n",
       "      <td>7.199098e+09</td>\n",
       "      <td>-0.141785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>glucose_minimal</td>\n",
       "      <td>0.210000</td>\n",
       "      <td>-6.324929e+08</td>\n",
       "      <td>4.438966e+09</td>\n",
       "      <td>-0.142487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>glucose_minimal</td>\n",
       "      <td>0.220000</td>\n",
       "      <td>-6.374814e+08</td>\n",
       "      <td>4.479156e+09</td>\n",
       "      <td>-0.142322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>glucose_minimal</td>\n",
       "      <td>0.260000</td>\n",
       "      <td>-6.634983e+08</td>\n",
       "      <td>4.644507e+09</td>\n",
       "      <td>-0.142857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>glucose_minimal</td>\n",
       "      <td>0.310000</td>\n",
       "      <td>-6.981347e+08</td>\n",
       "      <td>4.862400e+09</td>\n",
       "      <td>-0.143578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>glucose_minimal</td>\n",
       "      <td>0.360000</td>\n",
       "      <td>-7.349312e+08</td>\n",
       "      <td>5.084070e+09</td>\n",
       "      <td>-0.144556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>glucose_minimal</td>\n",
       "      <td>0.410000</td>\n",
       "      <td>-7.671303e+08</td>\n",
       "      <td>5.321234e+09</td>\n",
       "      <td>-0.144164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>glucose_minimal</td>\n",
       "      <td>0.460000</td>\n",
       "      <td>-8.052165e+08</td>\n",
       "      <td>5.566319e+09</td>\n",
       "      <td>-0.144659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>glucose_minimal</td>\n",
       "      <td>0.510000</td>\n",
       "      <td>-8.420826e+08</td>\n",
       "      <td>5.822963e+09</td>\n",
       "      <td>-0.144614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>rich</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>-6.597571e+08</td>\n",
       "      <td>4.600757e+09</td>\n",
       "      <td>-0.143402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>rich</td>\n",
       "      <td>0.350000</td>\n",
       "      <td>-7.293104e+08</td>\n",
       "      <td>5.043853e+09</td>\n",
       "      <td>-0.144594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>rich</td>\n",
       "      <td>0.420000</td>\n",
       "      <td>-7.833005e+08</td>\n",
       "      <td>5.374832e+09</td>\n",
       "      <td>-0.145735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>rich</td>\n",
       "      <td>0.450000</td>\n",
       "      <td>-8.031299e+08</td>\n",
       "      <td>5.523264e+09</td>\n",
       "      <td>-0.145409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>rich</td>\n",
       "      <td>0.530000</td>\n",
       "      <td>-8.651828e+08</td>\n",
       "      <td>5.938555e+09</td>\n",
       "      <td>-0.145689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>rich</td>\n",
       "      <td>0.550000</td>\n",
       "      <td>-8.820095e+08</td>\n",
       "      <td>6.045472e+09</td>\n",
       "      <td>-0.145896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>rich</td>\n",
       "      <td>0.630000</td>\n",
       "      <td>-9.499700e+08</td>\n",
       "      <td>6.501129e+09</td>\n",
       "      <td>-0.146124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>rich</td>\n",
       "      <td>0.650000</td>\n",
       "      <td>-9.652454e+08</td>\n",
       "      <td>6.617495e+09</td>\n",
       "      <td>-0.145863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>rich</td>\n",
       "      <td>0.730000</td>\n",
       "      <td>-1.042670e+09</td>\n",
       "      <td>7.150444e+09</td>\n",
       "      <td>-0.145819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>rich</td>\n",
       "      <td>0.740000</td>\n",
       "      <td>-1.049435e+09</td>\n",
       "      <td>7.207547e+09</td>\n",
       "      <td>-0.145602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>rich</td>\n",
       "      <td>0.780000</td>\n",
       "      <td>-1.085847e+09</td>\n",
       "      <td>7.451456e+09</td>\n",
       "      <td>-0.145723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>peebo_2015</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>rich</td>\n",
       "      <td>0.820000</td>\n",
       "      <td>-1.109044e+09</td>\n",
       "      <td>7.700418e+09</td>\n",
       "      <td>-0.144024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>42C</td>\n",
       "      <td>0.660000</td>\n",
       "      <td>-9.671966e+08</td>\n",
       "      <td>6.637538e+09</td>\n",
       "      <td>-0.145716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>acetate</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>-6.905831e+08</td>\n",
       "      <td>4.809245e+09</td>\n",
       "      <td>-0.143595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>chemostat_u0.12</td>\n",
       "      <td>0.120000</td>\n",
       "      <td>-5.857367e+08</td>\n",
       "      <td>4.090873e+09</td>\n",
       "      <td>-0.143181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>chemostat_u0.2</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>-6.341208e+08</td>\n",
       "      <td>4.398094e+09</td>\n",
       "      <td>-0.144181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>chemostat_u0.35</td>\n",
       "      <td>0.350000</td>\n",
       "      <td>-7.330251e+08</td>\n",
       "      <td>5.043504e+09</td>\n",
       "      <td>-0.145340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>chemostat_u0.5</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>-8.470172e+08</td>\n",
       "      <td>5.773070e+09</td>\n",
       "      <td>-0.146719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>fructose</td>\n",
       "      <td>0.650000</td>\n",
       "      <td>-9.750835e+08</td>\n",
       "      <td>6.597722e+09</td>\n",
       "      <td>-0.147791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>fumarate</td>\n",
       "      <td>0.420000</td>\n",
       "      <td>-7.775423e+08</td>\n",
       "      <td>5.362198e+09</td>\n",
       "      <td>-0.145004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>galactose</td>\n",
       "      <td>0.260000</td>\n",
       "      <td>-6.629211e+08</td>\n",
       "      <td>4.631581e+09</td>\n",
       "      <td>-0.143131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>glucosamine</td>\n",
       "      <td>0.460000</td>\n",
       "      <td>-8.076769e+08</td>\n",
       "      <td>5.547950e+09</td>\n",
       "      <td>-0.145581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>glucose</td>\n",
       "      <td>0.580000</td>\n",
       "      <td>-9.058484e+08</td>\n",
       "      <td>6.189567e+09</td>\n",
       "      <td>-0.146351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>glycerol</td>\n",
       "      <td>0.470000</td>\n",
       "      <td>-8.155418e+08</td>\n",
       "      <td>5.606386e+09</td>\n",
       "      <td>-0.145467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>glycerol_pAA</td>\n",
       "      <td>1.270000</td>\n",
       "      <td>-1.650778e+09</td>\n",
       "      <td>1.126306e+10</td>\n",
       "      <td>-0.146566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>lb_miller</td>\n",
       "      <td>1.900000</td>\n",
       "      <td>-2.823808e+09</td>\n",
       "      <td>1.922318e+10</td>\n",
       "      <td>-0.146896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>mannose</td>\n",
       "      <td>0.470000</td>\n",
       "      <td>-8.190262e+08</td>\n",
       "      <td>5.608490e+09</td>\n",
       "      <td>-0.146033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>osmotic_stress_glucose</td>\n",
       "      <td>0.550000</td>\n",
       "      <td>-8.677065e+08</td>\n",
       "      <td>6.007930e+09</td>\n",
       "      <td>-0.144427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>pH6</td>\n",
       "      <td>0.630000</td>\n",
       "      <td>-9.364280e+08</td>\n",
       "      <td>6.462866e+09</td>\n",
       "      <td>-0.144894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>pyruvate</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>-7.780363e+08</td>\n",
       "      <td>5.264003e+09</td>\n",
       "      <td>-0.147803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>succinate</td>\n",
       "      <td>0.440000</td>\n",
       "      <td>-7.971685e+08</td>\n",
       "      <td>5.459354e+09</td>\n",
       "      <td>-0.146019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>schmidt_2016</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>xylose</td>\n",
       "      <td>0.550000</td>\n",
       "      <td>-8.766555e+08</td>\n",
       "      <td>6.019325e+09</td>\n",
       "      <td>-0.145640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>valgepea_2013</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>glucose_minimal</td>\n",
       "      <td>0.110000</td>\n",
       "      <td>-5.761127e+08</td>\n",
       "      <td>3.947419e+09</td>\n",
       "      <td>-0.145947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>valgepea_2013</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>glucose_minimal</td>\n",
       "      <td>0.210000</td>\n",
       "      <td>-6.045871e+08</td>\n",
       "      <td>4.096512e+09</td>\n",
       "      <td>-0.147586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>valgepea_2013</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>glucose_minimal</td>\n",
       "      <td>0.310000</td>\n",
       "      <td>-6.741863e+08</td>\n",
       "      <td>4.544383e+09</td>\n",
       "      <td>-0.148356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>valgepea_2013</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>glucose_minimal</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>-7.389149e+08</td>\n",
       "      <td>4.976073e+09</td>\n",
       "      <td>-0.148494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>valgepea_2013</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>glucose_minimal</td>\n",
       "      <td>0.490000</td>\n",
       "      <td>-7.957861e+08</td>\n",
       "      <td>5.373643e+09</td>\n",
       "      <td>-0.148091</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          dataset   strain                         condition  growth_rate_hr  \\\n",
       "0         li_2014   MG1655                     MOPS complete        1.934364   \n",
       "1         li_2014   MG1655  MOPS complete without methionine        1.569390   \n",
       "2         li_2014   MG1655                      MOPS minimal        0.738700   \n",
       "3      peebo_2015  BW25113                   glucose_minimal        0.210000   \n",
       "4      peebo_2015  BW25113                   glucose_minimal        0.220000   \n",
       "5      peebo_2015  BW25113                   glucose_minimal        0.260000   \n",
       "6      peebo_2015  BW25113                   glucose_minimal        0.310000   \n",
       "7      peebo_2015  BW25113                   glucose_minimal        0.360000   \n",
       "8      peebo_2015  BW25113                   glucose_minimal        0.410000   \n",
       "9      peebo_2015  BW25113                   glucose_minimal        0.460000   \n",
       "10     peebo_2015  BW25113                   glucose_minimal        0.510000   \n",
       "11     peebo_2015  BW25113                              rich        0.250000   \n",
       "12     peebo_2015  BW25113                              rich        0.350000   \n",
       "13     peebo_2015  BW25113                              rich        0.420000   \n",
       "14     peebo_2015  BW25113                              rich        0.450000   \n",
       "15     peebo_2015  BW25113                              rich        0.530000   \n",
       "16     peebo_2015  BW25113                              rich        0.550000   \n",
       "17     peebo_2015  BW25113                              rich        0.630000   \n",
       "18     peebo_2015  BW25113                              rich        0.650000   \n",
       "19     peebo_2015  BW25113                              rich        0.730000   \n",
       "20     peebo_2015  BW25113                              rich        0.740000   \n",
       "21     peebo_2015  BW25113                              rich        0.780000   \n",
       "22     peebo_2015  BW25113                              rich        0.820000   \n",
       "23   schmidt_2016  BW25113                               42C        0.660000   \n",
       "24   schmidt_2016  BW25113                           acetate        0.300000   \n",
       "25   schmidt_2016  BW25113                   chemostat_u0.12        0.120000   \n",
       "26   schmidt_2016  BW25113                    chemostat_u0.2        0.200000   \n",
       "27   schmidt_2016  BW25113                   chemostat_u0.35        0.350000   \n",
       "28   schmidt_2016  BW25113                    chemostat_u0.5        0.500000   \n",
       "29   schmidt_2016  BW25113                          fructose        0.650000   \n",
       "30   schmidt_2016  BW25113                          fumarate        0.420000   \n",
       "31   schmidt_2016  BW25113                         galactose        0.260000   \n",
       "32   schmidt_2016  BW25113                       glucosamine        0.460000   \n",
       "33   schmidt_2016  BW25113                           glucose        0.580000   \n",
       "34   schmidt_2016  BW25113                          glycerol        0.470000   \n",
       "35   schmidt_2016  BW25113                      glycerol_pAA        1.270000   \n",
       "36   schmidt_2016  BW25113                         lb_miller        1.900000   \n",
       "37   schmidt_2016  BW25113                           mannose        0.470000   \n",
       "38   schmidt_2016  BW25113            osmotic_stress_glucose        0.550000   \n",
       "39   schmidt_2016  BW25113                               pH6        0.630000   \n",
       "40   schmidt_2016  BW25113                          pyruvate        0.400000   \n",
       "41   schmidt_2016  BW25113                         succinate        0.440000   \n",
       "42   schmidt_2016  BW25113                            xylose        0.550000   \n",
       "43  valgepea_2013   MG1655                   glucose_minimal        0.110000   \n",
       "44  valgepea_2013   MG1655                   glucose_minimal        0.210000   \n",
       "45  valgepea_2013   MG1655                   glucose_minimal        0.310000   \n",
       "46  valgepea_2013   MG1655                   glucose_minimal        0.400000   \n",
       "47  valgepea_2013   MG1655                   glucose_minimal        0.490000   \n",
       "\n",
       "        Ce_total      NC_total  proteome_NOSC  \n",
       "0  -2.873416e+09  1.979388e+10      -0.145167  \n",
       "1  -2.085635e+09  1.456026e+10      -0.143242  \n",
       "2  -1.020722e+09  7.199098e+09      -0.141785  \n",
       "3  -6.324929e+08  4.438966e+09      -0.142487  \n",
       "4  -6.374814e+08  4.479156e+09      -0.142322  \n",
       "5  -6.634983e+08  4.644507e+09      -0.142857  \n",
       "6  -6.981347e+08  4.862400e+09      -0.143578  \n",
       "7  -7.349312e+08  5.084070e+09      -0.144556  \n",
       "8  -7.671303e+08  5.321234e+09      -0.144164  \n",
       "9  -8.052165e+08  5.566319e+09      -0.144659  \n",
       "10 -8.420826e+08  5.822963e+09      -0.144614  \n",
       "11 -6.597571e+08  4.600757e+09      -0.143402  \n",
       "12 -7.293104e+08  5.043853e+09      -0.144594  \n",
       "13 -7.833005e+08  5.374832e+09      -0.145735  \n",
       "14 -8.031299e+08  5.523264e+09      -0.145409  \n",
       "15 -8.651828e+08  5.938555e+09      -0.145689  \n",
       "16 -8.820095e+08  6.045472e+09      -0.145896  \n",
       "17 -9.499700e+08  6.501129e+09      -0.146124  \n",
       "18 -9.652454e+08  6.617495e+09      -0.145863  \n",
       "19 -1.042670e+09  7.150444e+09      -0.145819  \n",
       "20 -1.049435e+09  7.207547e+09      -0.145602  \n",
       "21 -1.085847e+09  7.451456e+09      -0.145723  \n",
       "22 -1.109044e+09  7.700418e+09      -0.144024  \n",
       "23 -9.671966e+08  6.637538e+09      -0.145716  \n",
       "24 -6.905831e+08  4.809245e+09      -0.143595  \n",
       "25 -5.857367e+08  4.090873e+09      -0.143181  \n",
       "26 -6.341208e+08  4.398094e+09      -0.144181  \n",
       "27 -7.330251e+08  5.043504e+09      -0.145340  \n",
       "28 -8.470172e+08  5.773070e+09      -0.146719  \n",
       "29 -9.750835e+08  6.597722e+09      -0.147791  \n",
       "30 -7.775423e+08  5.362198e+09      -0.145004  \n",
       "31 -6.629211e+08  4.631581e+09      -0.143131  \n",
       "32 -8.076769e+08  5.547950e+09      -0.145581  \n",
       "33 -9.058484e+08  6.189567e+09      -0.146351  \n",
       "34 -8.155418e+08  5.606386e+09      -0.145467  \n",
       "35 -1.650778e+09  1.126306e+10      -0.146566  \n",
       "36 -2.823808e+09  1.922318e+10      -0.146896  \n",
       "37 -8.190262e+08  5.608490e+09      -0.146033  \n",
       "38 -8.677065e+08  6.007930e+09      -0.144427  \n",
       "39 -9.364280e+08  6.462866e+09      -0.144894  \n",
       "40 -7.780363e+08  5.264003e+09      -0.147803  \n",
       "41 -7.971685e+08  5.459354e+09      -0.146019  \n",
       "42 -8.766555e+08  6.019325e+09      -0.145640  \n",
       "43 -5.761127e+08  3.947419e+09      -0.145947  \n",
       "44 -6.045871e+08  4.096512e+09      -0.147586  \n",
       "45 -6.741863e+08  4.544383e+09      -0.148356  \n",
       "46 -7.389149e+08  4.976073e+09      -0.148494  \n",
       "47 -7.957861e+08  5.373643e+09      -0.148091  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteome_nosc_df = abund_df.groupby('dataset,strain,condition,growth_rate_hr'.split(',')).sum()\n",
    "proteome_nosc_df = proteome_nosc_df[['Ce_total', 'NC_total']].copy() \n",
    "proteome_nosc_df['proteome_NOSC'] = proteome_nosc_df.Ce_total / proteome_nosc_df.NC_total\n",
    "proteome_nosc_df = proteome_nosc_df.reset_index()\n",
    "proteome_nosc_df.to_csv('../data/proteomes/Coli/Chure_proteome_NOSC.csv', index=False)\n",
    "proteome_nosc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda.base",
   "language": "python",
   "name": "conda.base"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
