{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import re\n",
    "\n",
    "from matplotlib import pyplot as plt\n",
    "from collections import Counter"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyzing NOSC of the S. cerevisiae proteome\n",
    "Data from Xia et al. Nat Comms 2022. According to the \"description of supplementary files\" for that paper, the expression data are in units of molecules per cell. \n",
    "\n",
    "# Known issues\n",
    "Expression data contains ≈50 proteins with multiple majority IDs. This is currently resolved by adding fictional IDs with data for the first of those IDs. Mostly these are small sequence variants with almost no effects on NOSC. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>majority_protein_ids</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>S1</th>\n",
       "      <th>S2</th>\n",
       "      <th>S3</th>\n",
       "      <th>S4</th>\n",
       "      <th>S5</th>\n",
       "      <th>S6</th>\n",
       "      <th>S7</th>\n",
       "      <th>S8</th>\n",
       "      <th>...</th>\n",
       "      <th>S19</th>\n",
       "      <th>S20</th>\n",
       "      <th>S21</th>\n",
       "      <th>S22</th>\n",
       "      <th>S23</th>\n",
       "      <th>S24</th>\n",
       "      <th>S25</th>\n",
       "      <th>S26</th>\n",
       "      <th>S27</th>\n",
       "      <th>gene_function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P19097</td>\n",
       "      <td>FAS2</td>\n",
       "      <td>82452.270205</td>\n",
       "      <td>79779.523009</td>\n",
       "      <td>84618.632024</td>\n",
       "      <td>118520.675453</td>\n",
       "      <td>117488.905399</td>\n",
       "      <td>117159.998068</td>\n",
       "      <td>138977.483255</td>\n",
       "      <td>144138.512881</td>\n",
       "      <td>...</td>\n",
       "      <td>85063.315645</td>\n",
       "      <td>93305.693602</td>\n",
       "      <td>88758.854070</td>\n",
       "      <td>78471.524958</td>\n",
       "      <td>81948.073191</td>\n",
       "      <td>72495.589492</td>\n",
       "      <td>62546.099164</td>\n",
       "      <td>56921.725042</td>\n",
       "      <td>74664.084377</td>\n",
       "      <td>lipid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q00955</td>\n",
       "      <td>ACC1</td>\n",
       "      <td>80980.039401</td>\n",
       "      <td>79181.536246</td>\n",
       "      <td>83903.335061</td>\n",
       "      <td>86035.271225</td>\n",
       "      <td>79691.734225</td>\n",
       "      <td>85331.089655</td>\n",
       "      <td>67574.941552</td>\n",
       "      <td>70916.743805</td>\n",
       "      <td>...</td>\n",
       "      <td>50902.456412</td>\n",
       "      <td>56970.872752</td>\n",
       "      <td>53373.639893</td>\n",
       "      <td>50524.880588</td>\n",
       "      <td>51603.976968</td>\n",
       "      <td>46927.974480</td>\n",
       "      <td>40988.634801</td>\n",
       "      <td>36394.374344</td>\n",
       "      <td>48429.078610</td>\n",
       "      <td>lipid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P07259</td>\n",
       "      <td>URA2</td>\n",
       "      <td>39354.814940</td>\n",
       "      <td>38889.618852</td>\n",
       "      <td>40467.620523</td>\n",
       "      <td>42649.741464</td>\n",
       "      <td>41779.577050</td>\n",
       "      <td>41948.514658</td>\n",
       "      <td>50886.047984</td>\n",
       "      <td>52430.940434</td>\n",
       "      <td>...</td>\n",
       "      <td>42549.345866</td>\n",
       "      <td>52366.807193</td>\n",
       "      <td>44114.959866</td>\n",
       "      <td>52879.038353</td>\n",
       "      <td>53874.218583</td>\n",
       "      <td>50731.992629</td>\n",
       "      <td>44877.261859</td>\n",
       "      <td>41946.010033</td>\n",
       "      <td>59242.826490</td>\n",
       "      <td>Nucleotides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P07149</td>\n",
       "      <td>FAS1</td>\n",
       "      <td>70992.082651</td>\n",
       "      <td>68074.744978</td>\n",
       "      <td>71888.271167</td>\n",
       "      <td>99290.383630</td>\n",
       "      <td>98498.151949</td>\n",
       "      <td>98328.464667</td>\n",
       "      <td>116102.009908</td>\n",
       "      <td>121262.573677</td>\n",
       "      <td>...</td>\n",
       "      <td>70537.897812</td>\n",
       "      <td>75996.292451</td>\n",
       "      <td>72702.279944</td>\n",
       "      <td>65449.876876</td>\n",
       "      <td>68052.050691</td>\n",
       "      <td>58567.930810</td>\n",
       "      <td>51500.342385</td>\n",
       "      <td>47162.486880</td>\n",
       "      <td>59724.931330</td>\n",
       "      <td>lipid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P06105</td>\n",
       "      <td>SCP160</td>\n",
       "      <td>15866.563114</td>\n",
       "      <td>15589.045692</td>\n",
       "      <td>16002.005154</td>\n",
       "      <td>17532.347516</td>\n",
       "      <td>16867.256547</td>\n",
       "      <td>16872.126352</td>\n",
       "      <td>22262.699513</td>\n",
       "      <td>23154.853437</td>\n",
       "      <td>...</td>\n",
       "      <td>25762.256045</td>\n",
       "      <td>28125.209247</td>\n",
       "      <td>25498.954660</td>\n",
       "      <td>30447.872840</td>\n",
       "      <td>30838.126270</td>\n",
       "      <td>30499.993751</td>\n",
       "      <td>28210.792000</td>\n",
       "      <td>23627.567774</td>\n",
       "      <td>35529.349996</td>\n",
       "      <td>ER</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  majority_protein_ids gene_name            S1            S2            S3  \\\n",
       "0               P19097      FAS2  82452.270205  79779.523009  84618.632024   \n",
       "1               Q00955      ACC1  80980.039401  79181.536246  83903.335061   \n",
       "2               P07259      URA2  39354.814940  38889.618852  40467.620523   \n",
       "3               P07149      FAS1  70992.082651  68074.744978  71888.271167   \n",
       "4               P06105    SCP160  15866.563114  15589.045692  16002.005154   \n",
       "\n",
       "              S4             S5             S6             S7             S8  \\\n",
       "0  118520.675453  117488.905399  117159.998068  138977.483255  144138.512881   \n",
       "1   86035.271225   79691.734225   85331.089655   67574.941552   70916.743805   \n",
       "2   42649.741464   41779.577050   41948.514658   50886.047984   52430.940434   \n",
       "3   99290.383630   98498.151949   98328.464667  116102.009908  121262.573677   \n",
       "4   17532.347516   16867.256547   16872.126352   22262.699513   23154.853437   \n",
       "\n",
       "   ...           S19           S20           S21           S22           S23  \\\n",
       "0  ...  85063.315645  93305.693602  88758.854070  78471.524958  81948.073191   \n",
       "1  ...  50902.456412  56970.872752  53373.639893  50524.880588  51603.976968   \n",
       "2  ...  42549.345866  52366.807193  44114.959866  52879.038353  53874.218583   \n",
       "3  ...  70537.897812  75996.292451  72702.279944  65449.876876  68052.050691   \n",
       "4  ...  25762.256045  28125.209247  25498.954660  30447.872840  30838.126270   \n",
       "\n",
       "            S24           S25           S26           S27     gene_function  \n",
       "0  72495.589492  62546.099164  56921.725042  74664.084377  lipid metabolism  \n",
       "1  46927.974480  40988.634801  36394.374344  48429.078610  lipid metabolism  \n",
       "2  50731.992629  44877.261859  41946.010033  59242.826490       Nucleotides  \n",
       "3  58567.930810  51500.342385  47162.486880  59724.931330  lipid metabolism  \n",
       "4  30499.993751  28210.792000  23627.567774  35529.349996                ER  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_df = pd.read_excel('../data/proteomes/Scer/Xia_ScerCEN.PK.xlsx', sheet_name='samples', index_col=0)\n",
    "raw_abund_df = pd.read_excel('../data/proteomes/Scer/Xia_ScerCEN.PK.xlsx', sheet_name='data')\n",
    "nosc_df = pd.read_csv('../data/genomes/Scer/S288c/S288c_ref_prot_NOSC.csv')\n",
    "\n",
    "raw_abund_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing 45 ids\n",
      "Adding fictional IDs for those representing a mixture of isoforms\n",
      "Adding fictional protein for P0CX52:P0CX51 representing 2 isoforms\n",
      "Adding fictional protein for P0CX34:P0CX33 representing 2 isoforms\n",
      "Adding fictional protein for P0CX44:P0CX43 representing 2 isoforms\n",
      "Adding fictional protein for Q3E754:P0C0V8 representing 2 isoforms\n",
      "Adding fictional protein for P0CX83:P0CX82 representing 2 isoforms\n",
      "Adding fictional protein for P0CX38:P0CX37 representing 2 isoforms\n",
      "Adding fictional protein for P0CX48:P0CX47 representing 2 isoforms\n",
      "Adding fictional protein for P15565-2:P15565 representing 2 isoforms\n",
      "Adding fictional protein for P0CX26:P0CX25 representing 2 isoforms\n",
      "Adding fictional protein for P38431:P38431-2 representing 2 isoforms\n",
      "Adding fictional protein for P0CX85:P0CX84 representing 2 isoforms\n",
      "Adding fictional protein for P07806-2:P07806 representing 2 isoforms\n",
      "Adding fictional protein for P0CX50:P0CX49 representing 2 isoforms\n",
      "Adding fictional protein for P07884-2:P07884 representing 2 isoforms\n",
      "Adding fictional protein for P53174:P39551 representing 2 isoforms\n",
      "Adding fictional protein for P04912:P04911 representing 2 isoforms\n",
      "Adding fictional protein for Q3E757:P0C0W9 representing 2 isoforms\n",
      "Adding fictional protein for Q01532-2:Q01532 representing 2 isoforms\n",
      "Adding fictional protein for P04819-2:P04819 representing 2 isoforms\n",
      "Adding fictional protein for P41921-2:P41921 representing 2 isoforms\n",
      "Adding fictional protein for P00724-2:P00724 representing 2 isoforms\n",
      "Adding fictional protein for P0CX46:P0CX45 representing 2 isoforms\n",
      "Adding fictional protein for P0CX42:P0CX41 representing 2 isoforms\n",
      "Adding fictional protein for P53341:P38158 representing 2 isoforms\n",
      "Adding fictional protein for Q08911:P0CF35 representing 1 isoforms\n",
      "Adding fictional protein for P0CX24:P0CX23 representing 2 isoforms\n",
      "Adding fictional protein for P38711:P35997 representing 2 isoforms\n",
      "Adding fictional protein for P0CX28:P0CX27 representing 2 isoforms\n",
      "Adding fictional protein for P38784-2:P38784 representing 2 isoforms\n",
      "Adding fictional protein for Q3E7Y3:P0C0W1 representing 2 isoforms\n",
      "Adding fictional protein for P21269-3:P21269-2:P21269 representing 3 isoforms\n",
      "Adding fictional protein for P42222:P0CX11:P0CX10 representing 3 isoforms\n",
      "Adding fictional protein for P32796-2:P32796 representing 2 isoforms\n",
      "Adding fictional protein for P32831-2:P32831 representing 2 isoforms\n",
      "Adding fictional protein for Q01976-2:Q01976 representing 2 isoforms\n",
      "Adding fictional protein for P0CX36:P0CX35 representing 2 isoforms\n",
      "Adding fictional protein for P40825-2:P40825 representing 2 isoforms\n",
      "Adding fictional protein for Q08985:Q04898 representing 2 isoforms\n",
      "Adding fictional protein for P0CX40:P0CX39 representing 2 isoforms\n",
      "Adding fictional protein for P0CX56:P0CX55 representing 2 isoforms\n",
      "Adding fictional protein for P0C2H6:P0C2H7 representing 2 isoforms\n",
      "Adding fictional protein for P0CX32:P0CX31 representing 2 isoforms\n",
      "Adding fictional protein for P06778-3:P06778-2:P06778 representing 3 isoforms\n",
      "Adding fictional protein for P0CX30:P0CX29 representing 2 isoforms\n",
      "Adding fictional protein for P0CX54:P0CX53 representing 2 isoforms\n",
      "After update, missing 0 IDs\n"
     ]
    }
   ],
   "source": [
    "abund_ids = set(raw_abund_df.majority_protein_ids.values.tolist())\n",
    "cds_ids = set(nosc_df.primary_accession.values.tolist())\n",
    "\n",
    "# There are ≈50 entries where there was > 1 majority hit.\n",
    "# That is: the relevant peptides report on multiple proteins, often\n",
    "# because they are alternate translations of the same gene. \n",
    "missing_ids = abund_ids.difference(cds_ids)\n",
    "print('Missing {0} ids'.format(len(missing_ids)))\n",
    "shared_ids = abund_ids.intersection(cds_ids)\n",
    "\n",
    "# The missing IDs are mostly due to isoforms of proteins that differ slightly in sequence. \n",
    "# This code identifies the individual IDs and makes a fictional row that represents the average\n",
    "# of each of the isoforms.\n",
    "print('Adding fictional IDs for those representing a mixture of isoforms')\n",
    "lookup_table = nosc_df.set_index('primary_accession')\n",
    "fakes = []\n",
    "for my_id in missing_ids:\n",
    "    NCs = []\n",
    "    Ces = []\n",
    "    for x in my_id.split(':'):\n",
    "        if x in lookup_table.index:\n",
    "            row = lookup_table.loc[x]\n",
    "            NCs.append(row.NC)\n",
    "            Ces.append(row.Ce)\n",
    "            \n",
    "    if len(NCs) == 0:\n",
    "        continue\n",
    "    print('Adding fictional protein for {0} representing {1} isoforms'.format(\n",
    "        my_id, len(NCs)))\n",
    "    \n",
    "    NC = np.mean(NCs)\n",
    "    Ce = np.mean(Ces)\n",
    "    fake_protein = dict(primary_accession=my_id, NC=NC, Ce=Ce, NOSC=(Ce/NC))\n",
    "    fakes.append(fake_protein)\n",
    "    \n",
    "extended_nosc_df = pd.concat([nosc_df, pd.DataFrame(fakes)], ignore_index=True)\n",
    "\n",
    "# recheck which IDs are missing\n",
    "cds_ids = set(extended_nosc_df.primary_accession.values.tolist())\n",
    "missing_ids = abund_ids.difference(cds_ids)\n",
    "print('After update, missing {0} IDs'.format(len(missing_ids)))\n",
    "shared_ids = abund_ids.intersection(cds_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "Passing a set as an indexer is not supported. Use a list instead.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/h9/fvnb4xnn63qg71jx6ncvg62w0000gn/T/ipykernel_32745/4024974248.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Checking the percentage of unmapped genes.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mdata_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mraw_abund_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mmapped_sum\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mraw_abund_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'majority_protein_ids'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdata_cols\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mshared_ids\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mtotal\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mraw_abund_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdata_cols\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mpct_diff\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m100\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtotal\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mmapped_sum\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m/\u001b[0m\u001b[0mtotal\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/miniconda3/lib/python3.8/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1089\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mfinal\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1090\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__getitem__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1091\u001b[0;31m         \u001b[0mcheck_dict_or_set_indexers\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1092\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1093\u001b[0m             \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mis_iterator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/miniconda3/lib/python3.8/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36mcheck_dict_or_set_indexers\u001b[0;34m(key)\u001b[0m\n\u001b[1;32m   2616\u001b[0m         \u001b[0;32mand\u001b[0m \u001b[0many\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2617\u001b[0m     ):\n\u001b[0;32m-> 2618\u001b[0;31m         raise TypeError(\n\u001b[0m\u001b[1;32m   2619\u001b[0m             \u001b[0;34m\"Passing a set as an indexer is not supported. Use a list instead.\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2620\u001b[0m         )\n",
      "\u001b[0;31mTypeError\u001b[0m: Passing a set as an indexer is not supported. Use a list instead."
     ]
    }
   ],
   "source": [
    "# Checking the percentage of unmapped genes. \n",
    "data_cols = raw_abund_df.columns[2:-1]\n",
    "shared_ids_list = list(shared_ids)\n",
    "mapped_sum = raw_abund_df.set_index('majority_protein_ids')[data_cols].loc[shared_ids_list].sum()\n",
    "total = raw_abund_df[data_cols].sum()\n",
    "pct_diff = 100*(total - mapped_sum)/total\n",
    "\n",
    "# Now that we've handled the isoforms, we're counting all the expression data\n",
    "pct_diff.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NC_total</th>\n",
       "      <th>Ce_total</th>\n",
       "      <th>growth_rate_hr</th>\n",
       "      <th>growth_mode</th>\n",
       "      <th>proteome_NOSC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S1</th>\n",
       "      <td>1.244985e+11</td>\n",
       "      <td>-1.859931e+10</td>\n",
       "      <td>0.027</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.149394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S10</th>\n",
       "      <td>1.410082e+11</td>\n",
       "      <td>-2.146190e+10</td>\n",
       "      <td>0.152</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.152203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S11</th>\n",
       "      <td>1.529256e+11</td>\n",
       "      <td>-2.338499e+10</td>\n",
       "      <td>0.152</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.152917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S12</th>\n",
       "      <td>1.637752e+11</td>\n",
       "      <td>-2.509140e+10</td>\n",
       "      <td>0.152</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.153206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S13</th>\n",
       "      <td>8.950136e+10</td>\n",
       "      <td>-1.373303e+10</td>\n",
       "      <td>0.214</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.153439</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 NC_total      Ce_total  growth_rate_hr growth_mode  \\\n",
       "sample_name                                                           \n",
       "S1           1.244985e+11 -1.859931e+10           0.027   chemostat   \n",
       "S10          1.410082e+11 -2.146190e+10           0.152   chemostat   \n",
       "S11          1.529256e+11 -2.338499e+10           0.152   chemostat   \n",
       "S12          1.637752e+11 -2.509140e+10           0.152   chemostat   \n",
       "S13          8.950136e+10 -1.373303e+10           0.214   chemostat   \n",
       "\n",
       "             proteome_NOSC  \n",
       "sample_name                 \n",
       "S1               -0.149394  \n",
       "S10              -0.152203  \n",
       "S11              -0.152917  \n",
       "S12              -0.153206  \n",
       "S13              -0.153439  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reshape the data to long-form\n",
    "long_abund_df = raw_abund_df.drop('gene_function', axis=1).melt(\n",
    "    id_vars=['majority_protein_ids', 'gene_name'], var_name='sample_name',\n",
    "    value_name='copies_per_cell')\n",
    "\n",
    "growth_rates = samples_df.loc[long_abund_df.sample_name].mu_per_hr\n",
    "long_abund_df['growth_rate_hr'] = growth_rates.values\n",
    "\n",
    "# use the extended_nosc_df to calculate the condition-dependent proteome NOSC\n",
    "my_nosc_df = extended_nosc_df.set_index('primary_accession')\n",
    "NCs = my_nosc_df.loc[long_abund_df.majority_protein_ids].NC.values\n",
    "Ces = my_nosc_df.loc[long_abund_df.majority_protein_ids].Ce.values\n",
    "NOSCs = my_nosc_df.loc[long_abund_df.majority_protein_ids].NOSC.values\n",
    "long_abund_df['NC_per'] = NCs\n",
    "long_abund_df['Ce_per'] = Ces\n",
    "long_abund_df['NOSC'] = NOSCs\n",
    "long_abund_df['NC_total'] = long_abund_df.copies_per_cell.multiply(NCs)\n",
    "long_abund_df['Ce_total'] = long_abund_df.copies_per_cell.multiply(Ces)\n",
    "long_abund_df['dataset'] = 'xia_2022'\n",
    "long_abund_df['strain'] = 'CEN.PK113-7D'\n",
    "long_abund_df['species'] = 'S. cerevisiae'\n",
    "long_abund_df['organism_key'] = 'yeast'\n",
    "long_abund_df['condition'] = 'chemostat_u' + samples_df.loc[long_abund_df.sample_name].mu_per_hr.astype(str).values\n",
    "long_abund_df['fraction_transmembrane'] = my_nosc_df.loc[long_abund_df.majority_protein_ids].fraction_transmembrane.values\n",
    "long_abund_df['fraction_transmembrane_C'] = my_nosc_df.loc[long_abund_df.majority_protein_ids].fraction_transmembrane_C.values\n",
    "\n",
    "# Add annotation of the growth mode -- everything in this ref was done in chemostats\n",
    "long_abund_df['growth_mode'] = 'chemostat'\n",
    "# Add annotation of stress conds -- these are all glucose chemostat conds\n",
    "long_abund_df['stress'] = False\n",
    "\n",
    "# Save to CSV\n",
    "long_abund_df.to_csv('../data/proteomes/Scer/Xia_protein_measurements.csv', index=False)\n",
    "\n",
    "agg_dict = dict(NC_total='sum', Ce_total='sum', growth_rate_hr='first', growth_mode='first')\n",
    "sample_noscs = long_abund_df.groupby(['sample_name']).agg(agg_dict)\n",
    "sample_noscs['proteome_NOSC'] = sample_noscs.Ce_total / sample_noscs.NC_total\n",
    "\n",
    "# reset growth rates -- didn't want to sum them\n",
    "sample_noscs['growth_rate_hr'] = samples_df.loc[sample_noscs.index].mu_per_hr\n",
    "\n",
    "sample_noscs.to_csv('../data/proteomes/Scer/Xia_proteome_NOSC_full.csv', index=False)\n",
    "sample_noscs.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NC_total</th>\n",
       "      <th>Ce_total</th>\n",
       "      <th>growth_rate_hr</th>\n",
       "      <th>growth_mode</th>\n",
       "      <th>proteome_NOSC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S1</th>\n",
       "      <td>1.244985e+11</td>\n",
       "      <td>-1.859931e+10</td>\n",
       "      <td>0.027</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.149394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S10</th>\n",
       "      <td>1.410082e+11</td>\n",
       "      <td>-2.146190e+10</td>\n",
       "      <td>0.152</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.152203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S11</th>\n",
       "      <td>1.529256e+11</td>\n",
       "      <td>-2.338499e+10</td>\n",
       "      <td>0.152</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.152917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S12</th>\n",
       "      <td>1.637752e+11</td>\n",
       "      <td>-2.509140e+10</td>\n",
       "      <td>0.152</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.153206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S13</th>\n",
       "      <td>8.950136e+10</td>\n",
       "      <td>-1.373303e+10</td>\n",
       "      <td>0.214</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.153439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S14</th>\n",
       "      <td>1.254990e+11</td>\n",
       "      <td>-1.923848e+10</td>\n",
       "      <td>0.214</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.153296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S15</th>\n",
       "      <td>1.196692e+11</td>\n",
       "      <td>-1.846172e+10</td>\n",
       "      <td>0.214</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.154273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S16</th>\n",
       "      <td>1.240453e+11</td>\n",
       "      <td>-1.916775e+10</td>\n",
       "      <td>0.254</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.154522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S17</th>\n",
       "      <td>1.274194e+11</td>\n",
       "      <td>-1.966859e+10</td>\n",
       "      <td>0.254</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.154361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S18</th>\n",
       "      <td>1.300897e+11</td>\n",
       "      <td>-2.009779e+10</td>\n",
       "      <td>0.254</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.154492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S19</th>\n",
       "      <td>1.283529e+11</td>\n",
       "      <td>-1.990024e+10</td>\n",
       "      <td>0.284</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.155043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S2</th>\n",
       "      <td>1.244983e+11</td>\n",
       "      <td>-1.851412e+10</td>\n",
       "      <td>0.027</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.148710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S20</th>\n",
       "      <td>1.350025e+11</td>\n",
       "      <td>-2.089445e+10</td>\n",
       "      <td>0.284</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.154771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S21</th>\n",
       "      <td>1.304297e+11</td>\n",
       "      <td>-2.017187e+10</td>\n",
       "      <td>0.284</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.154657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S22</th>\n",
       "      <td>1.336799e+11</td>\n",
       "      <td>-2.075475e+10</td>\n",
       "      <td>0.334</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.155257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S23</th>\n",
       "      <td>1.343192e+11</td>\n",
       "      <td>-2.076958e+10</td>\n",
       "      <td>0.334</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.154629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S24</th>\n",
       "      <td>1.287077e+11</td>\n",
       "      <td>-1.992015e+10</td>\n",
       "      <td>0.334</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.154770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S25</th>\n",
       "      <td>1.186851e+11</td>\n",
       "      <td>-1.834838e+10</td>\n",
       "      <td>0.379</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.154597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S26</th>\n",
       "      <td>9.532132e+10</td>\n",
       "      <td>-1.476990e+10</td>\n",
       "      <td>0.379</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.154949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S27</th>\n",
       "      <td>1.419020e+11</td>\n",
       "      <td>-2.205304e+10</td>\n",
       "      <td>0.379</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.155410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S3</th>\n",
       "      <td>1.268499e+11</td>\n",
       "      <td>-1.888863e+10</td>\n",
       "      <td>0.027</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.148905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S4</th>\n",
       "      <td>1.311645e+11</td>\n",
       "      <td>-1.963912e+10</td>\n",
       "      <td>0.044</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.149729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S5</th>\n",
       "      <td>1.275098e+11</td>\n",
       "      <td>-1.913529e+10</td>\n",
       "      <td>0.044</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.150069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S6</th>\n",
       "      <td>1.297241e+11</td>\n",
       "      <td>-1.940327e+10</td>\n",
       "      <td>0.044</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.149573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S7</th>\n",
       "      <td>1.414456e+11</td>\n",
       "      <td>-2.145525e+10</td>\n",
       "      <td>0.102</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.151686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S8</th>\n",
       "      <td>1.436561e+11</td>\n",
       "      <td>-2.176747e+10</td>\n",
       "      <td>0.102</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.151525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S9</th>\n",
       "      <td>1.386144e+11</td>\n",
       "      <td>-2.097248e+10</td>\n",
       "      <td>0.102</td>\n",
       "      <td>chemostat</td>\n",
       "      <td>-0.151301</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 NC_total      Ce_total  growth_rate_hr growth_mode  \\\n",
       "sample_name                                                           \n",
       "S1           1.244985e+11 -1.859931e+10           0.027   chemostat   \n",
       "S10          1.410082e+11 -2.146190e+10           0.152   chemostat   \n",
       "S11          1.529256e+11 -2.338499e+10           0.152   chemostat   \n",
       "S12          1.637752e+11 -2.509140e+10           0.152   chemostat   \n",
       "S13          8.950136e+10 -1.373303e+10           0.214   chemostat   \n",
       "S14          1.254990e+11 -1.923848e+10           0.214   chemostat   \n",
       "S15          1.196692e+11 -1.846172e+10           0.214   chemostat   \n",
       "S16          1.240453e+11 -1.916775e+10           0.254   chemostat   \n",
       "S17          1.274194e+11 -1.966859e+10           0.254   chemostat   \n",
       "S18          1.300897e+11 -2.009779e+10           0.254   chemostat   \n",
       "S19          1.283529e+11 -1.990024e+10           0.284   chemostat   \n",
       "S2           1.244983e+11 -1.851412e+10           0.027   chemostat   \n",
       "S20          1.350025e+11 -2.089445e+10           0.284   chemostat   \n",
       "S21          1.304297e+11 -2.017187e+10           0.284   chemostat   \n",
       "S22          1.336799e+11 -2.075475e+10           0.334   chemostat   \n",
       "S23          1.343192e+11 -2.076958e+10           0.334   chemostat   \n",
       "S24          1.287077e+11 -1.992015e+10           0.334   chemostat   \n",
       "S25          1.186851e+11 -1.834838e+10           0.379   chemostat   \n",
       "S26          9.532132e+10 -1.476990e+10           0.379   chemostat   \n",
       "S27          1.419020e+11 -2.205304e+10           0.379   chemostat   \n",
       "S3           1.268499e+11 -1.888863e+10           0.027   chemostat   \n",
       "S4           1.311645e+11 -1.963912e+10           0.044   chemostat   \n",
       "S5           1.275098e+11 -1.913529e+10           0.044   chemostat   \n",
       "S6           1.297241e+11 -1.940327e+10           0.044   chemostat   \n",
       "S7           1.414456e+11 -2.145525e+10           0.102   chemostat   \n",
       "S8           1.436561e+11 -2.176747e+10           0.102   chemostat   \n",
       "S9           1.386144e+11 -2.097248e+10           0.102   chemostat   \n",
       "\n",
       "             proteome_NOSC  \n",
       "sample_name                 \n",
       "S1               -0.149394  \n",
       "S10              -0.152203  \n",
       "S11              -0.152917  \n",
       "S12              -0.153206  \n",
       "S13              -0.153439  \n",
       "S14              -0.153296  \n",
       "S15              -0.154273  \n",
       "S16              -0.154522  \n",
       "S17              -0.154361  \n",
       "S18              -0.154492  \n",
       "S19              -0.155043  \n",
       "S2               -0.148710  \n",
       "S20              -0.154771  \n",
       "S21              -0.154657  \n",
       "S22              -0.155257  \n",
       "S23              -0.154629  \n",
       "S24              -0.154770  \n",
       "S25              -0.154597  \n",
       "S26              -0.154949  \n",
       "S27              -0.155410  \n",
       "S3               -0.148905  \n",
       "S4               -0.149729  \n",
       "S5               -0.150069  \n",
       "S6               -0.149573  \n",
       "S7               -0.151686  \n",
       "S8               -0.151525  \n",
       "S9               -0.151301  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_noscs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NC_total</th>\n",
       "      <th>Ce_total</th>\n",
       "      <th>proteome_NOSC</th>\n",
       "      <th>sample_count</th>\n",
       "      <th>dataset</th>\n",
       "      <th>strain</th>\n",
       "      <th>condition</th>\n",
       "      <th>growth_mode</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>growth_rate_hr</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0.027</th>\n",
       "      <td>1.252822e+11</td>\n",
       "      <td>-1.866736e+10</td>\n",
       "      <td>-0.149003</td>\n",
       "      <td>3</td>\n",
       "      <td>xia_2022</td>\n",
       "      <td>CEN.PK113-7D</td>\n",
       "      <td>chemostat_u0.027</td>\n",
       "      <td>chemostat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.044</th>\n",
       "      <td>1.294661e+11</td>\n",
       "      <td>-1.939256e+10</td>\n",
       "      <td>-0.149791</td>\n",
       "      <td>3</td>\n",
       "      <td>xia_2022</td>\n",
       "      <td>CEN.PK113-7D</td>\n",
       "      <td>chemostat_u0.044</td>\n",
       "      <td>chemostat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.102</th>\n",
       "      <td>1.412387e+11</td>\n",
       "      <td>-2.139840e+10</td>\n",
       "      <td>-0.151504</td>\n",
       "      <td>3</td>\n",
       "      <td>xia_2022</td>\n",
       "      <td>CEN.PK113-7D</td>\n",
       "      <td>chemostat_u0.102</td>\n",
       "      <td>chemostat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.152</th>\n",
       "      <td>1.525697e+11</td>\n",
       "      <td>-2.331276e+10</td>\n",
       "      <td>-0.152776</td>\n",
       "      <td>3</td>\n",
       "      <td>xia_2022</td>\n",
       "      <td>CEN.PK113-7D</td>\n",
       "      <td>chemostat_u0.152</td>\n",
       "      <td>chemostat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.214</th>\n",
       "      <td>1.115565e+11</td>\n",
       "      <td>-1.714441e+10</td>\n",
       "      <td>-0.153669</td>\n",
       "      <td>3</td>\n",
       "      <td>xia_2022</td>\n",
       "      <td>CEN.PK113-7D</td>\n",
       "      <td>chemostat_u0.214</td>\n",
       "      <td>chemostat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.254</th>\n",
       "      <td>1.271848e+11</td>\n",
       "      <td>-1.964471e+10</td>\n",
       "      <td>-0.154458</td>\n",
       "      <td>3</td>\n",
       "      <td>xia_2022</td>\n",
       "      <td>CEN.PK113-7D</td>\n",
       "      <td>chemostat_u0.254</td>\n",
       "      <td>chemostat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.284</th>\n",
       "      <td>1.312617e+11</td>\n",
       "      <td>-2.032219e+10</td>\n",
       "      <td>-0.154824</td>\n",
       "      <td>3</td>\n",
       "      <td>xia_2022</td>\n",
       "      <td>CEN.PK113-7D</td>\n",
       "      <td>chemostat_u0.284</td>\n",
       "      <td>chemostat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.334</th>\n",
       "      <td>1.322356e+11</td>\n",
       "      <td>-2.048149e+10</td>\n",
       "      <td>-0.154885</td>\n",
       "      <td>3</td>\n",
       "      <td>xia_2022</td>\n",
       "      <td>CEN.PK113-7D</td>\n",
       "      <td>chemostat_u0.334</td>\n",
       "      <td>chemostat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.379</th>\n",
       "      <td>1.186361e+11</td>\n",
       "      <td>-1.839044e+10</td>\n",
       "      <td>-0.154985</td>\n",
       "      <td>3</td>\n",
       "      <td>xia_2022</td>\n",
       "      <td>CEN.PK113-7D</td>\n",
       "      <td>chemostat_u0.379</td>\n",
       "      <td>chemostat</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    NC_total      Ce_total  proteome_NOSC  sample_count  \\\n",
       "growth_rate_hr                                                            \n",
       "0.027           1.252822e+11 -1.866736e+10      -0.149003             3   \n",
       "0.044           1.294661e+11 -1.939256e+10      -0.149791             3   \n",
       "0.102           1.412387e+11 -2.139840e+10      -0.151504             3   \n",
       "0.152           1.525697e+11 -2.331276e+10      -0.152776             3   \n",
       "0.214           1.115565e+11 -1.714441e+10      -0.153669             3   \n",
       "0.254           1.271848e+11 -1.964471e+10      -0.154458             3   \n",
       "0.284           1.312617e+11 -2.032219e+10      -0.154824             3   \n",
       "0.334           1.322356e+11 -2.048149e+10      -0.154885             3   \n",
       "0.379           1.186361e+11 -1.839044e+10      -0.154985             3   \n",
       "\n",
       "                 dataset        strain         condition growth_mode  \n",
       "growth_rate_hr                                                        \n",
       "0.027           xia_2022  CEN.PK113-7D  chemostat_u0.027   chemostat  \n",
       "0.044           xia_2022  CEN.PK113-7D  chemostat_u0.044   chemostat  \n",
       "0.102           xia_2022  CEN.PK113-7D  chemostat_u0.102   chemostat  \n",
       "0.152           xia_2022  CEN.PK113-7D  chemostat_u0.152   chemostat  \n",
       "0.214           xia_2022  CEN.PK113-7D  chemostat_u0.214   chemostat  \n",
       "0.254           xia_2022  CEN.PK113-7D  chemostat_u0.254   chemostat  \n",
       "0.284           xia_2022  CEN.PK113-7D  chemostat_u0.284   chemostat  \n",
       "0.334           xia_2022  CEN.PK113-7D  chemostat_u0.334   chemostat  \n",
       "0.379           xia_2022  CEN.PK113-7D  chemostat_u0.379   chemostat  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Mean of S. cer data since the replicates are reported separately \n",
    "agg_dict = dict(NC_total='mean', Ce_total='mean', proteome_NOSC='mean', sample_name='count')\n",
    "sample_noscs_mean = sample_noscs.reset_index().groupby('growth_rate_hr').agg(agg_dict).rename(\n",
    "    columns=dict(sample_name='sample_count'))\n",
    "sample_noscs_mean['dataset'] = 'xia_2022'\n",
    "sample_noscs_mean['strain'] = 'CEN.PK113-7D'\n",
    "sample_noscs_mean['condition'] = 'chemostat_u' + sample_noscs_mean.index.astype(str)\n",
    "sample_noscs_mean['growth_mode'] = 'chemostat'\n",
    "\n",
    "sample_noscs_mean.to_csv('../data/proteomes/Scer/Xia_proteome_NOSC.csv', index=True)\n",
    "sample_noscs_mean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>M</th>\n",
       "      <th>T</th>\n",
       "      <th>G</th>\n",
       "      <th>F</th>\n",
       "      <th>K</th>\n",
       "      <th>V</th>\n",
       "      <th>S</th>\n",
       "      <th>Y</th>\n",
       "      <th>I</th>\n",
       "      <th>L</th>\n",
       "      <th>A</th>\n",
       "      <th>R</th>\n",
       "      <th>N</th>\n",
       "      <th>E</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>H</th>\n",
       "      <th>W</th>\n",
       "      <th>P</th>\n",
       "      <th>Q</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>primary_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P38088-2</th>\n",
       "      <td>12</td>\n",
       "      <td>30.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P07263-2</th>\n",
       "      <td>11</td>\n",
       "      <td>27.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P07806-2</th>\n",
       "      <td>15</td>\n",
       "      <td>58.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P15565-2</th>\n",
       "      <td>12</td>\n",
       "      <td>30.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P38784-2</th>\n",
       "      <td>4</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    M     T     G     F      K     V     S     Y     I     L  \\\n",
       "primary_accession                                                              \n",
       "P38088-2           12  30.0  41.0  33.0   60.0  51.0  39.0  18.0  37.0  56.0   \n",
       "P07263-2           11  27.0  40.0  20.0   52.0  33.0  40.0  17.0  36.0  42.0   \n",
       "P07806-2           15  58.0  59.0  38.0  108.0  68.0  63.0  34.0  74.0  88.0   \n",
       "P15565-2           12  30.0  32.0  17.0   53.0  35.0  41.0  23.0  30.0  44.0   \n",
       "P38784-2            4  11.0   9.0   3.0   19.0   8.0  13.0  10.0   9.0  19.0   \n",
       "\n",
       "                      A     R     N     E     C     D     H     W     P     Q  \n",
       "primary_accession                                                              \n",
       "P38088-2           49.0  36.0  23.0  61.0   6.0  49.0  13.0   7.0  30.0  16.0  \n",
       "P07263-2           52.0  20.0  15.0  40.0   5.0  35.0   6.0   4.0  15.0  16.0  \n",
       "P07806-2           71.0  38.0  44.0  92.0  11.0  69.0  22.0  24.0  51.0  31.0  \n",
       "P15565-2           37.0  24.0  42.0  38.0  11.0  23.0  12.0   5.0  30.0  15.0  \n",
       "P38784-2            4.0   6.0   9.0  17.0   1.0  11.0   1.0   1.0   4.0  17.0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# amino acid counts per protein\n",
    "aa_counts = [Counter(a) for a in nosc_df.aa_seq]\n",
    "aa_counts_df = pd.DataFrame(aa_counts, index=nosc_df.primary_accession).replace({np.NaN: 0})\n",
    "aa_counts_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        mu_per_hr\n",
      "sample           \n",
      "S25         0.379\n",
      "S26         0.379\n",
      "S27         0.379\n"
     ]
    }
   ],
   "source": [
    "mask = samples_df.mu_per_hr == samples_df.mu_per_hr.max()\n",
    "fast_samples = samples_df[mask].index\n",
    "print(samples_df[mask])\n",
    "mean_abund_fast_growth = raw_abund_df.set_index('majority_protein_ids')[fast_samples].mean(axis=1)\n",
    "\n",
    "overlapping_ids = set(mean_abund_fast_growth.index).intersection(aa_counts_df.index)\n",
    "tmp = aa_counts_df.loc[overlapping_ids].multiply(mean_abund_fast_growth.loc[overlapping_ids], axis=0).sum()\n",
    "expression_weighted_aas = tmp / tmp.sum()\n",
    "expression_weighted_aas.name = 'aa_freq'\n",
    "\n",
    "expression_weighted_aas.to_csv('../data/proteomes/Scer/Xia_mu0.379_expression_weighted_aa_freqs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda.base",
   "language": "python",
   "name": "conda.base"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
