{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import re\n",
    "\n",
    "from matplotlib import pyplot as plt"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyzing NOSC of the Cyanobacterial (PCC6803) proteome\n",
    "Data from Zavrel et al. Elife 2019\n",
    "\n",
    "# Known issues\n",
    "Using a reference genome for 6803 GT-S, but the data for GT-L. Mapping is by gene name, which is imperfect. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene names</th>\n",
       "      <th>Majority protein IDs</th>\n",
       "      <th>Entry name</th>\n",
       "      <th>C: Gene ontology (biological process)</th>\n",
       "      <th>C: Gene ontology (cellular component)</th>\n",
       "      <th>C: Gene ontology (molecular function)</th>\n",
       "      <th>C: Gene ontology (GO)</th>\n",
       "      <th>Length</th>\n",
       "      <th>Peptides</th>\n",
       "      <th>...</th>\n",
       "      <th>Q-value</th>\n",
       "      <th>Score</th>\n",
       "      <th>Intensity</th>\n",
       "      <th>MS/MS count</th>\n",
       "      <th>27.5</th>\n",
       "      <th>55</th>\n",
       "      <th>110</th>\n",
       "      <th>220</th>\n",
       "      <th>440</th>\n",
       "      <th>1100</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>majority_protein_ids</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P05429</th>\n",
       "      <td>Photosystem II CP47 reaction center protein (P...</td>\n",
       "      <td>psbB slr0906</td>\n",
       "      <td>P05429</td>\n",
       "      <td>PSBB_SYNY3</td>\n",
       "      <td>photosynthetic electron transport in photosyst...</td>\n",
       "      <td>integral component of membrane [GO:0016021];pl...</td>\n",
       "      <td>chlorophyll binding [GO:0016168];electron tran...</td>\n",
       "      <td>chlorophyll binding [GO:0016168];electron tran...</td>\n",
       "      <td>507</td>\n",
       "      <td>29</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>323.31</td>\n",
       "      <td>1.050000e+11</td>\n",
       "      <td>1058</td>\n",
       "      <td>69.73314</td>\n",
       "      <td>62.31298</td>\n",
       "      <td>52.86816</td>\n",
       "      <td>64.77500</td>\n",
       "      <td>54.62756</td>\n",
       "      <td>37.33008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P09190</th>\n",
       "      <td>Cytochrome b559 subunit alpha (PSII reaction c...</td>\n",
       "      <td>psbE ssr3451</td>\n",
       "      <td>P09190</td>\n",
       "      <td>PSBE_SYNY3</td>\n",
       "      <td>photosynthetic electron transport chain [GO:00...</td>\n",
       "      <td>integral component of membrane [GO:0016021];ph...</td>\n",
       "      <td>heme binding [GO:0020037];metal ion binding [G...</td>\n",
       "      <td>heme binding [GO:0020037];integral component o...</td>\n",
       "      <td>81</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>180.14</td>\n",
       "      <td>1.431100e+10</td>\n",
       "      <td>107</td>\n",
       "      <td>30.13294</td>\n",
       "      <td>24.67194</td>\n",
       "      <td>20.30978</td>\n",
       "      <td>22.72638</td>\n",
       "      <td>19.82798</td>\n",
       "      <td>15.74768</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                          Protein names  \\\n",
       "majority_protein_ids                                                      \n",
       "P05429                Photosystem II CP47 reaction center protein (P...   \n",
       "P09190                Cytochrome b559 subunit alpha (PSII reaction c...   \n",
       "\n",
       "                        Gene names Majority protein IDs  Entry name  \\\n",
       "majority_protein_ids                                                  \n",
       "P05429                psbB slr0906               P05429  PSBB_SYNY3   \n",
       "P09190                psbE ssr3451               P09190  PSBE_SYNY3   \n",
       "\n",
       "                                  C: Gene ontology (biological process)  \\\n",
       "majority_protein_ids                                                      \n",
       "P05429                photosynthetic electron transport in photosyst...   \n",
       "P09190                photosynthetic electron transport chain [GO:00...   \n",
       "\n",
       "                                  C: Gene ontology (cellular component)  \\\n",
       "majority_protein_ids                                                      \n",
       "P05429                integral component of membrane [GO:0016021];pl...   \n",
       "P09190                integral component of membrane [GO:0016021];ph...   \n",
       "\n",
       "                                  C: Gene ontology (molecular function)  \\\n",
       "majority_protein_ids                                                      \n",
       "P05429                chlorophyll binding [GO:0016168];electron tran...   \n",
       "P09190                heme binding [GO:0020037];metal ion binding [G...   \n",
       "\n",
       "                                                  C: Gene ontology (GO)  \\\n",
       "majority_protein_ids                                                      \n",
       "P05429                chlorophyll binding [GO:0016168];electron tran...   \n",
       "P09190                heme binding [GO:0020037];integral component o...   \n",
       "\n",
       "                      Length  Peptides  ...  Q-value   Score     Intensity  \\\n",
       "majority_protein_ids                    ...                                  \n",
       "P05429                   507        29  ...      0.0  323.31  1.050000e+11   \n",
       "P09190                    81         5  ...      0.0  180.14  1.431100e+10   \n",
       "\n",
       "                      MS/MS count      27.5        55       110       220  \\\n",
       "majority_protein_ids                                                        \n",
       "P05429                       1058  69.73314  62.31298  52.86816  64.77500   \n",
       "P09190                        107  30.13294  24.67194  20.30978  22.72638   \n",
       "\n",
       "                           440      1100  \n",
       "majority_protein_ids                      \n",
       "P05429                54.62756  37.33008  \n",
       "P09190                19.82798  15.74768  \n",
       "\n",
       "[2 rows x 24 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conds_df = pd.read_csv('../data/proteomes/Synechocystis/Zavrel_PCC6803_conditions.csv', dtype=dict(cond_id='str')).set_index('cond_id')\n",
    "nosc_df = pd.read_csv('../data/genomes/Synechocystis/PCC6803/PCC6803_ref_prot_NOSC.csv')\n",
    "\n",
    "# add a majority_protein_ids in a format matching the yeast data\n",
    "raw_abund_df = pd.read_csv('../data/proteomes/Synechocystis/Zavrel_PCC6803_proteome.csv')\n",
    "raw_abund_df['majority_protein_ids'] = [':'.join(ids.split(';')) for ids in raw_abund_df['Majority protein IDs']]\n",
    "raw_abund_df = raw_abund_df.set_index('majority_protein_ids')\n",
    "raw_abund_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'P74732', 'Q6ZEL1:Q6ZEQ3:P73620:P72822:Q6ZEP8:Q6ZEL2:Q6ZEK6', 'P73734', 'Q6ZEM5:Q6YRW8', 'P74195:P74627:P74059', 'Q6YRS8:P73838', 'Q6YRW7:Q6ZEM6', 'P74303', 'P73009:P74083', 'P73037'}\n",
      "10 not found\n",
      "Adding fictional IDs for those representing a mixture of isoforms\n",
      "Adding fictional protein for Q6ZEL1:Q6ZEQ3:P73620:P72822:Q6ZEP8:Q6ZEL2:Q6ZEK6 representing 7 isoforms\n",
      "Adding fictional protein for Q6ZEM5:Q6YRW8 representing 2 isoforms\n",
      "Adding fictional protein for P74195:P74627:P74059 representing 3 isoforms\n",
      "Adding fictional protein for Q6YRS8:P73838 representing 2 isoforms\n",
      "Adding fictional protein for Q6YRW7:Q6ZEM6 representing 2 isoforms\n",
      "Adding fictional protein for P73009:P74083 representing 2 isoforms\n",
      "After update, missing 4 IDs\n",
      "{'P74732', 'P73734', 'P73037', 'P74303'}\n"
     ]
    }
   ],
   "source": [
    "abund_ids = set(raw_abund_df.index.values.tolist())\n",
    "cds_ids = set(nosc_df.primary_accession.values.tolist())\n",
    "\n",
    "# There are â‰ˆ50 entries where there was > 1 majority hit.\n",
    "# That is: the relevant peptides report on multiple proteins, often\n",
    "# because they are alternate translations of the same gene. \n",
    "missing_ids = abund_ids.difference(cds_ids)\n",
    "print(missing_ids)\n",
    "print(len(missing_ids), 'not found')\n",
    "\n",
    "shared_ids = abund_ids.intersection(cds_ids)\n",
    "\n",
    "# The missing IDs are mostly due to isoforms of proteins that differ slightly in sequence. \n",
    "# This code identifies the individual IDs and makes a fictional row that represents the average\n",
    "# of each of the isoforms.\n",
    "print('Adding fictional IDs for those representing a mixture of isoforms')\n",
    "lookup_table = nosc_df.set_index('primary_accession')\n",
    "fakes = []\n",
    "for my_id in missing_ids:\n",
    "    NCs = []\n",
    "    Ces = []\n",
    "    for x in my_id.split(':'):\n",
    "        if x in lookup_table.index:\n",
    "            row = lookup_table.loc[x]\n",
    "            NCs.append(row.NC)\n",
    "            Ces.append(row.Ce)\n",
    "            \n",
    "    if len(NCs) == 0:\n",
    "        continue\n",
    "    print('Adding fictional protein for {0} representing {1} isoforms'.format(\n",
    "        my_id, len(NCs)))\n",
    "    \n",
    "    NC = np.mean(NCs)\n",
    "    Ce = np.mean(Ces)\n",
    "    fake_protein = dict(primary_accession=my_id, NC=NC, Ce=Ce, NOSC=(Ce/NC))\n",
    "    fakes.append(fake_protein)\n",
    "    \n",
    "extended_nosc_df = nosc_df.append(fakes, ignore_index=True)\n",
    "\n",
    "# recheck which IDs are missing\n",
    "cds_ids = set(extended_nosc_df.primary_accession.values.tolist())\n",
    "missing_ids = abund_ids.difference(cds_ids)\n",
    "print('After update, missing {0} IDs'.format(len(missing_ids)))\n",
    "print(missing_ids)\n",
    "shared_ids = abund_ids.intersection(cds_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27.5    0.048848\n",
       "55      0.049458\n",
       "110     0.052726\n",
       "220     0.051666\n",
       "440     0.055324\n",
       "1100    0.060204\n",
       "dtype: float64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Checking the percentage of unmapped genes. \n",
    "data_cols = ['27.5', '55', '110', '220', '440', '1100']\n",
    "mapped_sum = raw_abund_df[data_cols].loc[shared_ids].sum()\n",
    "total = raw_abund_df[data_cols].sum()\n",
    "pct_diff = 100*(total - mapped_sum)/total\n",
    "\n",
    "# Missing only â‰ˆ0.05% of the proteome due to mapping failure.\n",
    "# Should be OK to proceed. \n",
    "pct_diff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>majority_protein_ids</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>num_aas</th>\n",
       "      <th>mw_daltons</th>\n",
       "      <th>red_light_intensity_uE_m_s</th>\n",
       "      <th>copies_per_cell</th>\n",
       "      <th>NC_per</th>\n",
       "      <th>Ce_per</th>\n",
       "      <th>NC_total</th>\n",
       "      <th>Ce_total</th>\n",
       "      <th>NOSC</th>\n",
       "      <th>fraction_transmembrane</th>\n",
       "      <th>fraction_transmembrane_C</th>\n",
       "      <th>condition</th>\n",
       "      <th>growth_mode</th>\n",
       "      <th>stress</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P05429</td>\n",
       "      <td>psbB,slr0906</td>\n",
       "      <td>507</td>\n",
       "      <td>55902.0</td>\n",
       "      <td>27.5</td>\n",
       "      <td>69.733140</td>\n",
       "      <td>2589.0</td>\n",
       "      <td>-403.72</td>\n",
       "      <td>180539.099460</td>\n",
       "      <td>-28152.663281</td>\n",
       "      <td>-0.155937</td>\n",
       "      <td>0.177515</td>\n",
       "      <td>0.183855</td>\n",
       "      <td>photobio_27.5_uE</td>\n",
       "      <td>photobioreactor</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P09190</td>\n",
       "      <td>psbE,ssr3451</td>\n",
       "      <td>81</td>\n",
       "      <td>9448.5</td>\n",
       "      <td>27.5</td>\n",
       "      <td>30.132940</td>\n",
       "      <td>432.0</td>\n",
       "      <td>-63.67</td>\n",
       "      <td>13017.430080</td>\n",
       "      <td>-1918.564290</td>\n",
       "      <td>-0.147384</td>\n",
       "      <td>0.172840</td>\n",
       "      <td>0.180556</td>\n",
       "      <td>photobio_27.5_uE</td>\n",
       "      <td>photobioreactor</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P09192</td>\n",
       "      <td>psbD,sll0849</td>\n",
       "      <td>352</td>\n",
       "      <td>39492.0</td>\n",
       "      <td>27.5</td>\n",
       "      <td>51.210580</td>\n",
       "      <td>1856.0</td>\n",
       "      <td>-358.02</td>\n",
       "      <td>95046.836480</td>\n",
       "      <td>-18334.411852</td>\n",
       "      <td>-0.192899</td>\n",
       "      <td>0.241477</td>\n",
       "      <td>0.235991</td>\n",
       "      <td>photobio_27.5_uE</td>\n",
       "      <td>photobioreactor</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P09193</td>\n",
       "      <td>psbC,sll0851</td>\n",
       "      <td>460</td>\n",
       "      <td>50302.0</td>\n",
       "      <td>27.5</td>\n",
       "      <td>60.468680</td>\n",
       "      <td>2337.0</td>\n",
       "      <td>-450.17</td>\n",
       "      <td>141315.305160</td>\n",
       "      <td>-27221.185676</td>\n",
       "      <td>-0.192627</td>\n",
       "      <td>0.195652</td>\n",
       "      <td>0.204536</td>\n",
       "      <td>photobio_27.5_uE</td>\n",
       "      <td>photobioreactor</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P0DJF8</td>\n",
       "      <td>mtnP,sll0135</td>\n",
       "      <td>326</td>\n",
       "      <td>35318.0</td>\n",
       "      <td>27.5</td>\n",
       "      <td>1.255867</td>\n",
       "      <td>1587.0</td>\n",
       "      <td>-289.36</td>\n",
       "      <td>1993.061564</td>\n",
       "      <td>-363.397791</td>\n",
       "      <td>-0.182331</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>photobio_27.5_uE</td>\n",
       "      <td>photobioreactor</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  majority_protein_ids     gene_name  num_aas  mw_daltons  \\\n",
       "0               P05429  psbB,slr0906      507     55902.0   \n",
       "1               P09190  psbE,ssr3451       81      9448.5   \n",
       "2               P09192  psbD,sll0849      352     39492.0   \n",
       "3               P09193  psbC,sll0851      460     50302.0   \n",
       "4               P0DJF8  mtnP,sll0135      326     35318.0   \n",
       "\n",
       "  red_light_intensity_uE_m_s  copies_per_cell  NC_per  Ce_per       NC_total  \\\n",
       "0                       27.5        69.733140  2589.0 -403.72  180539.099460   \n",
       "1                       27.5        30.132940   432.0  -63.67   13017.430080   \n",
       "2                       27.5        51.210580  1856.0 -358.02   95046.836480   \n",
       "3                       27.5        60.468680  2337.0 -450.17  141315.305160   \n",
       "4                       27.5         1.255867  1587.0 -289.36    1993.061564   \n",
       "\n",
       "       Ce_total      NOSC  fraction_transmembrane  fraction_transmembrane_C  \\\n",
       "0 -28152.663281 -0.155937                0.177515                  0.183855   \n",
       "1  -1918.564290 -0.147384                0.172840                  0.180556   \n",
       "2 -18334.411852 -0.192899                0.241477                  0.235991   \n",
       "3 -27221.185676 -0.192627                0.195652                  0.204536   \n",
       "4   -363.397791 -0.182331                0.000000                  0.000000   \n",
       "\n",
       "          condition      growth_mode  stress  \n",
       "0  photobio_27.5_uE  photobioreactor   False  \n",
       "1  photobio_27.5_uE  photobioreactor   False  \n",
       "2  photobio_27.5_uE  photobioreactor   False  \n",
       "3  photobio_27.5_uE  photobioreactor   False  \n",
       "4  photobio_27.5_uE  photobioreactor   False  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_cols = ['27.5', '55', '110', '220', '440', '1100']\n",
    "long_abund_df = raw_abund_df.reset_index().melt(\n",
    "    id_vars=['majority_protein_ids', 'Gene names', 'Length', 'Mol. weight [kDa]'],\n",
    "    value_vars=data_cols,\n",
    "    var_name='red_light_intensity_uE_m_s',\n",
    "    value_name='copies_per_cell').rename(\n",
    "    columns={'Gene names': 'gene_name',\n",
    "             'Mol. weight [kDa]': 'mw_daltons',\n",
    "             'Length': 'num_aas'})\n",
    "\n",
    "# use the extended_nosc_df to calculate the condition-dependent proteome NOSC\n",
    "# first need to add a few empty rows for which we don't have NOSC data for some reason.\n",
    "# TODO: figure out why not? we already know they account for very little expression (above).\n",
    "missing_ids = set(long_abund_df.majority_protein_ids).difference(\n",
    "    extended_nosc_df.primary_accession)\n",
    "empty_rows = pd.DataFrame(dict(primary_accession=list(missing_ids)))\n",
    "my_nosc_df = pd.concat([extended_nosc_df, empty_rows]).set_index('primary_accession')\n",
    "\n",
    "NCs = my_nosc_df.loc[long_abund_df.majority_protein_ids].NC.values\n",
    "Ces = my_nosc_df.loc[long_abund_df.majority_protein_ids].Ce.values\n",
    "NOSCs = my_nosc_df.loc[long_abund_df.majority_protein_ids].NOSC.values\n",
    "ftm = my_nosc_df.loc[long_abund_df.majority_protein_ids].fraction_transmembrane.values\n",
    "ftmC = my_nosc_df.loc[long_abund_df.majority_protein_ids].fraction_transmembrane_C.values\n",
    "long_abund_df['NC_per'] = NCs\n",
    "long_abund_df['Ce_per'] = Ces\n",
    "long_abund_df['NC_total'] = long_abund_df.copies_per_cell.multiply(NCs)\n",
    "long_abund_df['Ce_total'] = long_abund_df.copies_per_cell.multiply(Ces)\n",
    "long_abund_df['NOSC'] = NOSCs \n",
    "long_abund_df['fraction_transmembrane'] = ftm \n",
    "long_abund_df['fraction_transmembrane_C'] = ftmC \n",
    "long_abund_df['condition'] = (\n",
    "    'photobio_' + long_abund_df.red_light_intensity_uE_m_s.astype(str).values + '_uE')\n",
    "\n",
    "# Add annotation of the growth mode -- here photobioreactors in turbidostat mode\n",
    "long_abund_df['growth_mode'] = 'photobioreactor'\n",
    "# Add annotation of stress conds -- these are all growth in variable light conds\n",
    "long_abund_df['stress'] = False\n",
    "\n",
    "# convert to daltons\n",
    "long_abund_df.mw_daltons *= 1000\n",
    "# commas for multiple gene names\n",
    "long_abund_df.gene_name = long_abund_df.gene_name.str.replace(' ', ',')\n",
    "long_abund_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>red_light_intensity_uE_m_s</th>\n",
       "      <th>growth_rate_hr</th>\n",
       "      <th>spare_media_flow_rate_hr</th>\n",
       "      <th>dry_weight_mg_L</th>\n",
       "      <th>proteome_NOSC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cond_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27.5</th>\n",
       "      <td>27.5</td>\n",
       "      <td>0.025</td>\n",
       "      <td>0.028</td>\n",
       "      <td>132</td>\n",
       "      <td>-0.132964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>55.0</td>\n",
       "      <td>0.039</td>\n",
       "      <td>0.034</td>\n",
       "      <td>166</td>\n",
       "      <td>-0.134464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>110.0</td>\n",
       "      <td>0.059</td>\n",
       "      <td>0.048</td>\n",
       "      <td>173</td>\n",
       "      <td>-0.135985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>220.0</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0.070</td>\n",
       "      <td>220</td>\n",
       "      <td>-0.136507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>440</th>\n",
       "      <td>440.0</td>\n",
       "      <td>0.104</td>\n",
       "      <td>0.121</td>\n",
       "      <td>221</td>\n",
       "      <td>-0.136293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>660</th>\n",
       "      <td>660.0</td>\n",
       "      <td>0.104</td>\n",
       "      <td>0.118</td>\n",
       "      <td>260</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>880</th>\n",
       "      <td>880.0</td>\n",
       "      <td>0.990</td>\n",
       "      <td>0.089</td>\n",
       "      <td>256</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1100</th>\n",
       "      <td>1100.0</td>\n",
       "      <td>0.930</td>\n",
       "      <td>0.086</td>\n",
       "      <td>231</td>\n",
       "      <td>-0.139379</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         red_light_intensity_uE_m_s  growth_rate_hr  spare_media_flow_rate_hr  \\\n",
       "cond_id                                                                         \n",
       "27.5                           27.5           0.025                     0.028   \n",
       "55                             55.0           0.039                     0.034   \n",
       "110                           110.0           0.059                     0.048   \n",
       "220                           220.0           0.081                     0.070   \n",
       "440                           440.0           0.104                     0.121   \n",
       "660                           660.0           0.104                     0.118   \n",
       "880                           880.0           0.990                     0.089   \n",
       "1100                         1100.0           0.930                     0.086   \n",
       "\n",
       "         dry_weight_mg_L  proteome_NOSC  \n",
       "cond_id                                  \n",
       "27.5                 132      -0.132964  \n",
       "55                   166      -0.134464  \n",
       "110                  173      -0.135985  \n",
       "220                  220      -0.136507  \n",
       "440                  221      -0.136293  \n",
       "660                  260            NaN  \n",
       "880                  256            NaN  \n",
       "1100                 231      -0.139379  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Data re in fg/ul, convert to proteins/ul\n",
    "# TODO: save this unit conversion\n",
    "data_cols = ['27.5', '55', '110', '220', '440', '1100']\n",
    "# fg/mol = kg/mol * 1e18 fg/kg\n",
    "mws_fg_mol = raw_abund_df.loc[shared_ids]['Mol. weight [kDa]'] * 1e18 \n",
    "# copies/ul = 6.02e23 copies/mol * fg/ul / (fg/mol)\n",
    "copies_per_ul = raw_abund_df[data_cols].loc[shared_ids].multiply(6.02e23/mws_fg_mol, axis=0)\n",
    "\n",
    "Ces = my_nosc_df.loc[shared_ids].Ce.values\n",
    "NCs = my_nosc_df.loc[shared_ids].NC.values\n",
    "Ce_total = copies_per_ul.multiply(Ces, axis=0)\n",
    "NC_total = copies_per_ul.multiply(NCs, axis=0)\n",
    "\n",
    "cond_NOSC = Ce_total.sum()/NC_total.sum()\n",
    "conds_df['proteome_NOSC'] = cond_NOSC\n",
    "conds_df.to_csv('../data/proteomes/Synechocystis/Zavrel_proteome_NOSC.csv')\n",
    "\n",
    "conds_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Saving the full dataset with per-row annotations\n",
    "tmp = long_abund_df.copy()\n",
    "\n",
    "tmp['dataset'] = 'zavrel_2019'\n",
    "tmp['strain'] = 'PCC6803'\n",
    "tmp['species'] = 'Synechocystis sp.'\n",
    "tmp['organism_key'] = 'PCC6803'\n",
    "tmp['growth_rate_hr'] = conds_df.loc[tmp.red_light_intensity_uE_m_s].growth_rate_hr.values\n",
    "\n",
    "tmp.to_csv('../data/proteomes/Synechocystis/Zavrel_protein_measurements.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>majority_protein_ids</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>num_aas</th>\n",
       "      <th>mw_daltons</th>\n",
       "      <th>red_light_intensity_uE_m_s</th>\n",
       "      <th>copies_per_cell</th>\n",
       "      <th>NC_per</th>\n",
       "      <th>Ce_per</th>\n",
       "      <th>NC_total</th>\n",
       "      <th>Ce_total</th>\n",
       "      <th>...</th>\n",
       "      <th>fraction_transmembrane</th>\n",
       "      <th>fraction_transmembrane_C</th>\n",
       "      <th>condition</th>\n",
       "      <th>growth_mode</th>\n",
       "      <th>stress</th>\n",
       "      <th>dataset</th>\n",
       "      <th>strain</th>\n",
       "      <th>species</th>\n",
       "      <th>organism_key</th>\n",
       "      <th>growth_rate_hr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P05429</td>\n",
       "      <td>psbB,slr0906</td>\n",
       "      <td>507</td>\n",
       "      <td>55902.0</td>\n",
       "      <td>27.5</td>\n",
       "      <td>69.733140</td>\n",
       "      <td>2589.0</td>\n",
       "      <td>-403.72</td>\n",
       "      <td>180539.099460</td>\n",
       "      <td>-28152.663281</td>\n",
       "      <td>...</td>\n",
       "      <td>0.177515</td>\n",
       "      <td>0.183855</td>\n",
       "      <td>photobio_27.5_uE</td>\n",
       "      <td>photobioreactor</td>\n",
       "      <td>False</td>\n",
       "      <td>zavrel_2019</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>Synechocystis sp.</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>0.025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P09190</td>\n",
       "      <td>psbE,ssr3451</td>\n",
       "      <td>81</td>\n",
       "      <td>9448.5</td>\n",
       "      <td>27.5</td>\n",
       "      <td>30.132940</td>\n",
       "      <td>432.0</td>\n",
       "      <td>-63.67</td>\n",
       "      <td>13017.430080</td>\n",
       "      <td>-1918.564290</td>\n",
       "      <td>...</td>\n",
       "      <td>0.172840</td>\n",
       "      <td>0.180556</td>\n",
       "      <td>photobio_27.5_uE</td>\n",
       "      <td>photobioreactor</td>\n",
       "      <td>False</td>\n",
       "      <td>zavrel_2019</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>Synechocystis sp.</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>0.025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P09192</td>\n",
       "      <td>psbD,sll0849</td>\n",
       "      <td>352</td>\n",
       "      <td>39492.0</td>\n",
       "      <td>27.5</td>\n",
       "      <td>51.210580</td>\n",
       "      <td>1856.0</td>\n",
       "      <td>-358.02</td>\n",
       "      <td>95046.836480</td>\n",
       "      <td>-18334.411852</td>\n",
       "      <td>...</td>\n",
       "      <td>0.241477</td>\n",
       "      <td>0.235991</td>\n",
       "      <td>photobio_27.5_uE</td>\n",
       "      <td>photobioreactor</td>\n",
       "      <td>False</td>\n",
       "      <td>zavrel_2019</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>Synechocystis sp.</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>0.025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P09193</td>\n",
       "      <td>psbC,sll0851</td>\n",
       "      <td>460</td>\n",
       "      <td>50302.0</td>\n",
       "      <td>27.5</td>\n",
       "      <td>60.468680</td>\n",
       "      <td>2337.0</td>\n",
       "      <td>-450.17</td>\n",
       "      <td>141315.305160</td>\n",
       "      <td>-27221.185676</td>\n",
       "      <td>...</td>\n",
       "      <td>0.195652</td>\n",
       "      <td>0.204536</td>\n",
       "      <td>photobio_27.5_uE</td>\n",
       "      <td>photobioreactor</td>\n",
       "      <td>False</td>\n",
       "      <td>zavrel_2019</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>Synechocystis sp.</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>0.025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P0DJF8</td>\n",
       "      <td>mtnP,sll0135</td>\n",
       "      <td>326</td>\n",
       "      <td>35318.0</td>\n",
       "      <td>27.5</td>\n",
       "      <td>1.255867</td>\n",
       "      <td>1587.0</td>\n",
       "      <td>-289.36</td>\n",
       "      <td>1993.061564</td>\n",
       "      <td>-363.397791</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>photobio_27.5_uE</td>\n",
       "      <td>photobioreactor</td>\n",
       "      <td>False</td>\n",
       "      <td>zavrel_2019</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>Synechocystis sp.</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>0.025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10099</th>\n",
       "      <td>Q6ZEU6</td>\n",
       "      <td>sll5034</td>\n",
       "      <td>207</td>\n",
       "      <td>23200.0</td>\n",
       "      <td>1100</td>\n",
       "      <td>0.321208</td>\n",
       "      <td>1019.0</td>\n",
       "      <td>-133.93</td>\n",
       "      <td>327.310870</td>\n",
       "      <td>-43.019377</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>photobio_1100_uE</td>\n",
       "      <td>photobioreactor</td>\n",
       "      <td>False</td>\n",
       "      <td>zavrel_2019</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>Synechocystis sp.</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>0.930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10100</th>\n",
       "      <td>Q79EE4</td>\n",
       "      <td>ggtA</td>\n",
       "      <td>363</td>\n",
       "      <td>40755.0</td>\n",
       "      <td>1100</td>\n",
       "      <td>0.166779</td>\n",
       "      <td>1812.0</td>\n",
       "      <td>-301.54</td>\n",
       "      <td>302.202714</td>\n",
       "      <td>-50.290401</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>photobio_1100_uE</td>\n",
       "      <td>photobioreactor</td>\n",
       "      <td>False</td>\n",
       "      <td>zavrel_2019</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>Synechocystis sp.</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>0.930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10101</th>\n",
       "      <td>Q79EE5</td>\n",
       "      <td>slr1588</td>\n",
       "      <td>421</td>\n",
       "      <td>47088.0</td>\n",
       "      <td>1100</td>\n",
       "      <td>0.147759</td>\n",
       "      <td>2114.0</td>\n",
       "      <td>-383.21</td>\n",
       "      <td>312.362526</td>\n",
       "      <td>-56.622726</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>photobio_1100_uE</td>\n",
       "      <td>photobioreactor</td>\n",
       "      <td>False</td>\n",
       "      <td>zavrel_2019</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>Synechocystis sp.</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>0.930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10102</th>\n",
       "      <td>Q79EF1</td>\n",
       "      <td>desB</td>\n",
       "      <td>359</td>\n",
       "      <td>41918.0</td>\n",
       "      <td>1100</td>\n",
       "      <td>0.161586</td>\n",
       "      <td>1989.0</td>\n",
       "      <td>-361.97</td>\n",
       "      <td>321.394594</td>\n",
       "      <td>-58.489292</td>\n",
       "      <td>...</td>\n",
       "      <td>0.264624</td>\n",
       "      <td>0.284062</td>\n",
       "      <td>photobio_1100_uE</td>\n",
       "      <td>photobioreactor</td>\n",
       "      <td>False</td>\n",
       "      <td>zavrel_2019</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>Synechocystis sp.</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>0.930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10103</th>\n",
       "      <td>Q79EF9</td>\n",
       "      <td>mntC</td>\n",
       "      <td>330</td>\n",
       "      <td>36075.0</td>\n",
       "      <td>1100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1603.0</td>\n",
       "      <td>-205.43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>photobio_1100_uE</td>\n",
       "      <td>photobioreactor</td>\n",
       "      <td>False</td>\n",
       "      <td>zavrel_2019</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>Synechocystis sp.</td>\n",
       "      <td>PCC6803</td>\n",
       "      <td>0.930</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10104 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      majority_protein_ids     gene_name  num_aas  mw_daltons  \\\n",
       "0                   P05429  psbB,slr0906      507     55902.0   \n",
       "1                   P09190  psbE,ssr3451       81      9448.5   \n",
       "2                   P09192  psbD,sll0849      352     39492.0   \n",
       "3                   P09193  psbC,sll0851      460     50302.0   \n",
       "4                   P0DJF8  mtnP,sll0135      326     35318.0   \n",
       "...                    ...           ...      ...         ...   \n",
       "10099               Q6ZEU6       sll5034      207     23200.0   \n",
       "10100               Q79EE4          ggtA      363     40755.0   \n",
       "10101               Q79EE5       slr1588      421     47088.0   \n",
       "10102               Q79EF1          desB      359     41918.0   \n",
       "10103               Q79EF9          mntC      330     36075.0   \n",
       "\n",
       "      red_light_intensity_uE_m_s  copies_per_cell  NC_per  Ce_per  \\\n",
       "0                           27.5        69.733140  2589.0 -403.72   \n",
       "1                           27.5        30.132940   432.0  -63.67   \n",
       "2                           27.5        51.210580  1856.0 -358.02   \n",
       "3                           27.5        60.468680  2337.0 -450.17   \n",
       "4                           27.5         1.255867  1587.0 -289.36   \n",
       "...                          ...              ...     ...     ...   \n",
       "10099                       1100         0.321208  1019.0 -133.93   \n",
       "10100                       1100         0.166779  1812.0 -301.54   \n",
       "10101                       1100         0.147759  2114.0 -383.21   \n",
       "10102                       1100         0.161586  1989.0 -361.97   \n",
       "10103                       1100              NaN  1603.0 -205.43   \n",
       "\n",
       "            NC_total      Ce_total  ...  fraction_transmembrane  \\\n",
       "0      180539.099460 -28152.663281  ...                0.177515   \n",
       "1       13017.430080  -1918.564290  ...                0.172840   \n",
       "2       95046.836480 -18334.411852  ...                0.241477   \n",
       "3      141315.305160 -27221.185676  ...                0.195652   \n",
       "4        1993.061564   -363.397791  ...                0.000000   \n",
       "...              ...           ...  ...                     ...   \n",
       "10099     327.310870    -43.019377  ...                0.000000   \n",
       "10100     302.202714    -50.290401  ...                0.000000   \n",
       "10101     312.362526    -56.622726  ...                0.000000   \n",
       "10102     321.394594    -58.489292  ...                0.264624   \n",
       "10103            NaN           NaN  ...                0.000000   \n",
       "\n",
       "       fraction_transmembrane_C         condition      growth_mode stress  \\\n",
       "0                      0.183855  photobio_27.5_uE  photobioreactor  False   \n",
       "1                      0.180556  photobio_27.5_uE  photobioreactor  False   \n",
       "2                      0.235991  photobio_27.5_uE  photobioreactor  False   \n",
       "3                      0.204536  photobio_27.5_uE  photobioreactor  False   \n",
       "4                      0.000000  photobio_27.5_uE  photobioreactor  False   \n",
       "...                         ...               ...              ...    ...   \n",
       "10099                  0.000000  photobio_1100_uE  photobioreactor  False   \n",
       "10100                  0.000000  photobio_1100_uE  photobioreactor  False   \n",
       "10101                  0.000000  photobio_1100_uE  photobioreactor  False   \n",
       "10102                  0.284062  photobio_1100_uE  photobioreactor  False   \n",
       "10103                  0.000000  photobio_1100_uE  photobioreactor  False   \n",
       "\n",
       "           dataset   strain            species organism_key growth_rate_hr  \n",
       "0      zavrel_2019  PCC6803  Synechocystis sp.      PCC6803          0.025  \n",
       "1      zavrel_2019  PCC6803  Synechocystis sp.      PCC6803          0.025  \n",
       "2      zavrel_2019  PCC6803  Synechocystis sp.      PCC6803          0.025  \n",
       "3      zavrel_2019  PCC6803  Synechocystis sp.      PCC6803          0.025  \n",
       "4      zavrel_2019  PCC6803  Synechocystis sp.      PCC6803          0.025  \n",
       "...            ...      ...                ...          ...            ...  \n",
       "10099  zavrel_2019  PCC6803  Synechocystis sp.      PCC6803          0.930  \n",
       "10100  zavrel_2019  PCC6803  Synechocystis sp.      PCC6803          0.930  \n",
       "10101  zavrel_2019  PCC6803  Synechocystis sp.      PCC6803          0.930  \n",
       "10102  zavrel_2019  PCC6803  Synechocystis sp.      PCC6803          0.930  \n",
       "10103  zavrel_2019  PCC6803  Synechocystis sp.      PCC6803          0.930  \n",
       "\n",
       "[10104 rows x 21 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda.base",
   "language": "python",
   "name": "conda.base"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
