{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pingouin as pg\n",
    "import seaborn as sns\n",
    "\n",
    "from scipy.stats import linregress, gmean, pearsonr\n",
    "from sklearn.metrics import r2_score\n",
    "from tqdm import tqdm\n",
    "from Bio import SeqIO\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.colors import LogNorm\n",
    "from os import path\n",
    "#from seq_util import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>PF00380.20</th>\n",
       "      <th>PF00410.20</th>\n",
       "      <th>PF00466.21</th>\n",
       "      <th>PF01025.20</th>\n",
       "      <th>PF02576.18</th>\n",
       "      <th>PF03726.15</th>\n",
       "      <th>TIGR00006</th>\n",
       "      <th>TIGR00019</th>\n",
       "      <th>TIGR00020</th>\n",
       "      <th>...</th>\n",
       "      <th>TIGR02432</th>\n",
       "      <th>TIGR02729</th>\n",
       "      <th>TIGR03263</th>\n",
       "      <th>TIGR03594</th>\n",
       "      <th>TIGR03625</th>\n",
       "      <th>TIGR03632</th>\n",
       "      <th>TIGR03654</th>\n",
       "      <th>TIGR03723</th>\n",
       "      <th>TIGR03725</th>\n",
       "      <th>TIGR03953</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trimmed_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GB_GCA_000007325</th>\n",
       "      <td>GB_GCA_000007325.1</td>\n",
       "      <td>-0.148160</td>\n",
       "      <td>-0.224515</td>\n",
       "      <td>-0.271998</td>\n",
       "      <td>-0.122832</td>\n",
       "      <td>-0.222500</td>\n",
       "      <td>-0.195491</td>\n",
       "      <td>-0.189573</td>\n",
       "      <td>-0.152311</td>\n",
       "      <td>-0.116115</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.239489</td>\n",
       "      <td>-0.215168</td>\n",
       "      <td>-0.138229</td>\n",
       "      <td>-0.214850</td>\n",
       "      <td>-0.219437</td>\n",
       "      <td>-0.161122</td>\n",
       "      <td>-0.178828</td>\n",
       "      <td>-0.207375</td>\n",
       "      <td>-0.232777</td>\n",
       "      <td>-0.213205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_000008885</th>\n",
       "      <td>GB_GCA_000008885.1</td>\n",
       "      <td>-0.225820</td>\n",
       "      <td>-0.247991</td>\n",
       "      <td>-0.235449</td>\n",
       "      <td>-0.223839</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.212832</td>\n",
       "      <td>-0.275949</td>\n",
       "      <td>-0.212414</td>\n",
       "      <td>-0.160727</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.280723</td>\n",
       "      <td>-0.219518</td>\n",
       "      <td>-0.278769</td>\n",
       "      <td>-0.307224</td>\n",
       "      <td>-0.212944</td>\n",
       "      <td>-0.103641</td>\n",
       "      <td>-0.312473</td>\n",
       "      <td>-0.254505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.265804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_000009845</th>\n",
       "      <td>GB_GCA_000009845.1</td>\n",
       "      <td>-0.210141</td>\n",
       "      <td>-0.304301</td>\n",
       "      <td>-0.323290</td>\n",
       "      <td>-0.130158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.202842</td>\n",
       "      <td>-0.193418</td>\n",
       "      <td>-0.160336</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.243972</td>\n",
       "      <td>-0.205122</td>\n",
       "      <td>-0.180596</td>\n",
       "      <td>-0.214562</td>\n",
       "      <td>-0.172616</td>\n",
       "      <td>-0.165642</td>\n",
       "      <td>-0.229449</td>\n",
       "      <td>-0.217595</td>\n",
       "      <td>-0.288547</td>\n",
       "      <td>-0.173903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_000010085</th>\n",
       "      <td>GB_GCA_000010085.1</td>\n",
       "      <td>-0.097661</td>\n",
       "      <td>-0.237436</td>\n",
       "      <td>-0.161992</td>\n",
       "      <td>-0.060054</td>\n",
       "      <td>-0.167924</td>\n",
       "      <td>-0.131780</td>\n",
       "      <td>-0.111915</td>\n",
       "      <td>-0.063734</td>\n",
       "      <td>-0.064409</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.078462</td>\n",
       "      <td>-0.093497</td>\n",
       "      <td>-0.132986</td>\n",
       "      <td>-0.143900</td>\n",
       "      <td>-0.152612</td>\n",
       "      <td>-0.093533</td>\n",
       "      <td>-0.164809</td>\n",
       "      <td>-0.124286</td>\n",
       "      <td>-0.153650</td>\n",
       "      <td>-0.178778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_000010565</th>\n",
       "      <td>GB_GCA_000010565.1</td>\n",
       "      <td>-0.187030</td>\n",
       "      <td>-0.251788</td>\n",
       "      <td>-0.224343</td>\n",
       "      <td>-0.079801</td>\n",
       "      <td>-0.200000</td>\n",
       "      <td>-0.177482</td>\n",
       "      <td>-0.149577</td>\n",
       "      <td>-0.113914</td>\n",
       "      <td>-0.096177</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.179268</td>\n",
       "      <td>-0.101351</td>\n",
       "      <td>-0.181850</td>\n",
       "      <td>-0.177694</td>\n",
       "      <td>-0.190099</td>\n",
       "      <td>-0.100300</td>\n",
       "      <td>-0.222171</td>\n",
       "      <td>-0.170978</td>\n",
       "      <td>-0.203634</td>\n",
       "      <td>-0.148263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_907165215</th>\n",
       "      <td>RS_GCF_907165215.1</td>\n",
       "      <td>-0.168351</td>\n",
       "      <td>-0.201494</td>\n",
       "      <td>-0.224994</td>\n",
       "      <td>0.007326</td>\n",
       "      <td>-0.164994</td>\n",
       "      <td>-0.175330</td>\n",
       "      <td>-0.101722</td>\n",
       "      <td>-0.107393</td>\n",
       "      <td>-0.080776</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.094517</td>\n",
       "      <td>-0.119690</td>\n",
       "      <td>-0.146439</td>\n",
       "      <td>-0.160860</td>\n",
       "      <td>-0.166670</td>\n",
       "      <td>-0.079130</td>\n",
       "      <td>-0.202686</td>\n",
       "      <td>-0.165731</td>\n",
       "      <td>-0.105502</td>\n",
       "      <td>-0.180360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_910593725</th>\n",
       "      <td>RS_GCF_910593725.1</td>\n",
       "      <td>-0.135314</td>\n",
       "      <td>-0.215554</td>\n",
       "      <td>-0.229864</td>\n",
       "      <td>-0.090785</td>\n",
       "      <td>-0.128396</td>\n",
       "      <td>-0.183764</td>\n",
       "      <td>-0.143342</td>\n",
       "      <td>-0.056712</td>\n",
       "      <td>-0.057451</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.149315</td>\n",
       "      <td>-0.126020</td>\n",
       "      <td>-0.135354</td>\n",
       "      <td>-0.156069</td>\n",
       "      <td>-0.162119</td>\n",
       "      <td>-0.094318</td>\n",
       "      <td>-0.218050</td>\n",
       "      <td>-0.177884</td>\n",
       "      <td>-0.155941</td>\n",
       "      <td>-0.194961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_910593785</th>\n",
       "      <td>RS_GCF_910593785.1</td>\n",
       "      <td>-0.127126</td>\n",
       "      <td>-0.160112</td>\n",
       "      <td>-0.166364</td>\n",
       "      <td>-0.010832</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.149092</td>\n",
       "      <td>-0.114223</td>\n",
       "      <td>-0.054536</td>\n",
       "      <td>-0.100068</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.014610</td>\n",
       "      <td>-0.111930</td>\n",
       "      <td>-0.085451</td>\n",
       "      <td>-0.119093</td>\n",
       "      <td>-0.104254</td>\n",
       "      <td>-0.086850</td>\n",
       "      <td>-0.194778</td>\n",
       "      <td>-0.170642</td>\n",
       "      <td>-0.117131</td>\n",
       "      <td>-0.125924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_910593845</th>\n",
       "      <td>RS_GCF_910593845.1</td>\n",
       "      <td>-0.187393</td>\n",
       "      <td>-0.211284</td>\n",
       "      <td>-0.211267</td>\n",
       "      <td>-0.045940</td>\n",
       "      <td>-0.171825</td>\n",
       "      <td>-0.180315</td>\n",
       "      <td>-0.118257</td>\n",
       "      <td>-0.070155</td>\n",
       "      <td>-0.099315</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.115815</td>\n",
       "      <td>-0.124364</td>\n",
       "      <td>-0.157902</td>\n",
       "      <td>-0.183341</td>\n",
       "      <td>-0.184423</td>\n",
       "      <td>-0.102591</td>\n",
       "      <td>-0.189493</td>\n",
       "      <td>-0.174194</td>\n",
       "      <td>-0.109978</td>\n",
       "      <td>-0.217005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_910594985</th>\n",
       "      <td>RS_GCF_910594985.1</td>\n",
       "      <td>-0.187393</td>\n",
       "      <td>-0.213689</td>\n",
       "      <td>-0.194117</td>\n",
       "      <td>-0.030070</td>\n",
       "      <td>-0.184480</td>\n",
       "      <td>-0.182145</td>\n",
       "      <td>-0.126617</td>\n",
       "      <td>-0.075900</td>\n",
       "      <td>-0.095851</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.131454</td>\n",
       "      <td>-0.116531</td>\n",
       "      <td>-0.172154</td>\n",
       "      <td>-0.174063</td>\n",
       "      <td>-0.176663</td>\n",
       "      <td>-0.102591</td>\n",
       "      <td>-0.188467</td>\n",
       "      <td>-0.172784</td>\n",
       "      <td>-0.126123</td>\n",
       "      <td>-0.215229</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62291 rows × 121 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            accession  PF00380.20  PF00410.20  PF00466.21  \\\n",
       "trimmed_accession                                                           \n",
       "GB_GCA_000007325   GB_GCA_000007325.1   -0.148160   -0.224515   -0.271998   \n",
       "GB_GCA_000008885   GB_GCA_000008885.1   -0.225820   -0.247991   -0.235449   \n",
       "GB_GCA_000009845   GB_GCA_000009845.1   -0.210141   -0.304301   -0.323290   \n",
       "GB_GCA_000010085   GB_GCA_000010085.1   -0.097661   -0.237436   -0.161992   \n",
       "GB_GCA_000010565   GB_GCA_000010565.1   -0.187030   -0.251788   -0.224343   \n",
       "...                               ...         ...         ...         ...   \n",
       "RS_GCF_907165215   RS_GCF_907165215.1   -0.168351   -0.201494   -0.224994   \n",
       "RS_GCF_910593725   RS_GCF_910593725.1   -0.135314   -0.215554   -0.229864   \n",
       "RS_GCF_910593785   RS_GCF_910593785.1   -0.127126   -0.160112   -0.166364   \n",
       "RS_GCF_910593845   RS_GCF_910593845.1   -0.187393   -0.211284   -0.211267   \n",
       "RS_GCF_910594985   RS_GCF_910594985.1   -0.187393   -0.213689   -0.194117   \n",
       "\n",
       "                   PF01025.20  PF02576.18  PF03726.15  TIGR00006  TIGR00019  \\\n",
       "trimmed_accession                                                             \n",
       "GB_GCA_000007325    -0.122832   -0.222500   -0.195491  -0.189573  -0.152311   \n",
       "GB_GCA_000008885    -0.223839         NaN   -0.212832  -0.275949  -0.212414   \n",
       "GB_GCA_000009845    -0.130158         NaN         NaN  -0.202842  -0.193418   \n",
       "GB_GCA_000010085    -0.060054   -0.167924   -0.131780  -0.111915  -0.063734   \n",
       "GB_GCA_000010565    -0.079801   -0.200000   -0.177482  -0.149577  -0.113914   \n",
       "...                       ...         ...         ...        ...        ...   \n",
       "RS_GCF_907165215     0.007326   -0.164994   -0.175330  -0.101722  -0.107393   \n",
       "RS_GCF_910593725    -0.090785   -0.128396   -0.183764  -0.143342  -0.056712   \n",
       "RS_GCF_910593785    -0.010832         NaN   -0.149092  -0.114223  -0.054536   \n",
       "RS_GCF_910593845    -0.045940   -0.171825   -0.180315  -0.118257  -0.070155   \n",
       "RS_GCF_910594985    -0.030070   -0.184480   -0.182145  -0.126617  -0.075900   \n",
       "\n",
       "                   TIGR00020  ...  TIGR02432  TIGR02729  TIGR03263  TIGR03594  \\\n",
       "trimmed_accession             ...                                               \n",
       "GB_GCA_000007325   -0.116115  ...  -0.239489  -0.215168  -0.138229  -0.214850   \n",
       "GB_GCA_000008885   -0.160727  ...  -0.280723  -0.219518  -0.278769  -0.307224   \n",
       "GB_GCA_000009845   -0.160336  ...  -0.243972  -0.205122  -0.180596  -0.214562   \n",
       "GB_GCA_000010085   -0.064409  ...  -0.078462  -0.093497  -0.132986  -0.143900   \n",
       "GB_GCA_000010565   -0.096177  ...  -0.179268  -0.101351  -0.181850  -0.177694   \n",
       "...                      ...  ...        ...        ...        ...        ...   \n",
       "RS_GCF_907165215   -0.080776  ...  -0.094517  -0.119690  -0.146439  -0.160860   \n",
       "RS_GCF_910593725   -0.057451  ...  -0.149315  -0.126020  -0.135354  -0.156069   \n",
       "RS_GCF_910593785   -0.100068  ...  -0.014610  -0.111930  -0.085451  -0.119093   \n",
       "RS_GCF_910593845   -0.099315  ...  -0.115815  -0.124364  -0.157902  -0.183341   \n",
       "RS_GCF_910594985   -0.095851  ...  -0.131454  -0.116531  -0.172154  -0.174063   \n",
       "\n",
       "                   TIGR03625  TIGR03632  TIGR03654  TIGR03723  TIGR03725  \\\n",
       "trimmed_accession                                                          \n",
       "GB_GCA_000007325   -0.219437  -0.161122  -0.178828  -0.207375  -0.232777   \n",
       "GB_GCA_000008885   -0.212944  -0.103641  -0.312473  -0.254505        NaN   \n",
       "GB_GCA_000009845   -0.172616  -0.165642  -0.229449  -0.217595  -0.288547   \n",
       "GB_GCA_000010085   -0.152612  -0.093533  -0.164809  -0.124286  -0.153650   \n",
       "GB_GCA_000010565   -0.190099  -0.100300  -0.222171  -0.170978  -0.203634   \n",
       "...                      ...        ...        ...        ...        ...   \n",
       "RS_GCF_907165215   -0.166670  -0.079130  -0.202686  -0.165731  -0.105502   \n",
       "RS_GCF_910593725   -0.162119  -0.094318  -0.218050  -0.177884  -0.155941   \n",
       "RS_GCF_910593785   -0.104254  -0.086850  -0.194778  -0.170642  -0.117131   \n",
       "RS_GCF_910593845   -0.184423  -0.102591  -0.189493  -0.174194  -0.109978   \n",
       "RS_GCF_910594985   -0.176663  -0.102591  -0.188467  -0.172784  -0.126123   \n",
       "\n",
       "                   TIGR03953  \n",
       "trimmed_accession             \n",
       "GB_GCA_000007325   -0.213205  \n",
       "GB_GCA_000008885   -0.265804  \n",
       "GB_GCA_000009845   -0.173903  \n",
       "GB_GCA_000010085   -0.178778  \n",
       "GB_GCA_000010565   -0.148263  \n",
       "...                      ...  \n",
       "RS_GCF_907165215   -0.180360  \n",
       "RS_GCF_910593725   -0.194961  \n",
       "RS_GCF_910593785   -0.125924  \n",
       "RS_GCF_910593845   -0.217005  \n",
       "RS_GCF_910594985   -0.215229  \n",
       "\n",
       "[62291 rows x 121 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the NOSC values for bac120 as a wide-form matrix\n",
    "bac120_nosc_df = pd.read_csv('../data/gtdb/r207/bac120_nosc_vals_wide.csv', index_col=0)\n",
    "\n",
    "# Trim the \".1\" from the end of the accession numbers so they match other GTDB data\n",
    "trimmed_ids = [x.split('.')[0] for x in bac120_nosc_df.index.tolist()]\n",
    "bac120_nosc_df['trimmed_accession'] = trimmed_ids\n",
    "\n",
    "# Reset the index to the trimmed accession number\n",
    "bac120_nosc_df = bac120_nosc_df.reset_index().set_index('trimmed_accession')\n",
    "bac120_nosc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_avg_NOSC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genome_accession</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RS_GCF_000971475</th>\n",
       "      <td>-0.159401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_003693265</th>\n",
       "      <td>-0.119059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_018969935</th>\n",
       "      <td>-0.152630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_016790845</th>\n",
       "      <td>-0.158100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_900321855</th>\n",
       "      <td>-0.175164</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  genome_avg_NOSC\n",
       "genome_accession                 \n",
       "RS_GCF_000971475        -0.159401\n",
       "RS_GCF_003693265        -0.119059\n",
       "GB_GCA_018969935        -0.152630\n",
       "GB_GCA_016790845        -0.158100\n",
       "GB_GCA_900321855        -0.175164"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the annotated bac120 metadata\n",
    "bac120_meta_df = pd.read_csv('../data/gtdb/r207/bac120_msa_marker_info_r207_annot.csv',\n",
    "                             index_col=0)\n",
    "\n",
    "# Read the genome average NOSC values for representative genomes\n",
    "reps_genome_nosc_df = pd.read_csv('../data/gtdb/r207/genome_average_nosc.csv', index_col=0)\n",
    "reps_genome_nosc_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>PF00380.20</th>\n",
       "      <th>PF00410.20</th>\n",
       "      <th>PF00466.21</th>\n",
       "      <th>PF01025.20</th>\n",
       "      <th>PF02576.18</th>\n",
       "      <th>PF03726.15</th>\n",
       "      <th>TIGR00006</th>\n",
       "      <th>TIGR00019</th>\n",
       "      <th>TIGR00020</th>\n",
       "      <th>...</th>\n",
       "      <th>TIGR02729</th>\n",
       "      <th>TIGR03263</th>\n",
       "      <th>TIGR03594</th>\n",
       "      <th>TIGR03625</th>\n",
       "      <th>TIGR03632</th>\n",
       "      <th>TIGR03654</th>\n",
       "      <th>TIGR03723</th>\n",
       "      <th>TIGR03725</th>\n",
       "      <th>TIGR03953</th>\n",
       "      <th>genome_avg_NOSC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GB_GCA_000007325</th>\n",
       "      <td>GB_GCA_000007325.1</td>\n",
       "      <td>-0.148160</td>\n",
       "      <td>-0.224515</td>\n",
       "      <td>-0.271998</td>\n",
       "      <td>-0.122832</td>\n",
       "      <td>-0.222500</td>\n",
       "      <td>-0.195491</td>\n",
       "      <td>-0.189573</td>\n",
       "      <td>-0.152311</td>\n",
       "      <td>-0.116115</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.215168</td>\n",
       "      <td>-0.138229</td>\n",
       "      <td>-0.214850</td>\n",
       "      <td>-0.219437</td>\n",
       "      <td>-0.161122</td>\n",
       "      <td>-0.178828</td>\n",
       "      <td>-0.207375</td>\n",
       "      <td>-0.232777</td>\n",
       "      <td>-0.213205</td>\n",
       "      <td>-0.223900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_000008885</th>\n",
       "      <td>GB_GCA_000008885.1</td>\n",
       "      <td>-0.225820</td>\n",
       "      <td>-0.247991</td>\n",
       "      <td>-0.235449</td>\n",
       "      <td>-0.223839</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.212832</td>\n",
       "      <td>-0.275949</td>\n",
       "      <td>-0.212414</td>\n",
       "      <td>-0.160727</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.219518</td>\n",
       "      <td>-0.278769</td>\n",
       "      <td>-0.307224</td>\n",
       "      <td>-0.212944</td>\n",
       "      <td>-0.103641</td>\n",
       "      <td>-0.312473</td>\n",
       "      <td>-0.254505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.265804</td>\n",
       "      <td>-0.258189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_000009845</th>\n",
       "      <td>GB_GCA_000009845.1</td>\n",
       "      <td>-0.210141</td>\n",
       "      <td>-0.304301</td>\n",
       "      <td>-0.323290</td>\n",
       "      <td>-0.130158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.202842</td>\n",
       "      <td>-0.193418</td>\n",
       "      <td>-0.160336</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.205122</td>\n",
       "      <td>-0.180596</td>\n",
       "      <td>-0.214562</td>\n",
       "      <td>-0.172616</td>\n",
       "      <td>-0.165642</td>\n",
       "      <td>-0.229449</td>\n",
       "      <td>-0.217595</td>\n",
       "      <td>-0.288547</td>\n",
       "      <td>-0.173903</td>\n",
       "      <td>-0.225228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_000010085</th>\n",
       "      <td>GB_GCA_000010085.1</td>\n",
       "      <td>-0.097661</td>\n",
       "      <td>-0.237436</td>\n",
       "      <td>-0.161992</td>\n",
       "      <td>-0.060054</td>\n",
       "      <td>-0.167924</td>\n",
       "      <td>-0.131780</td>\n",
       "      <td>-0.111915</td>\n",
       "      <td>-0.063734</td>\n",
       "      <td>-0.064409</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.093497</td>\n",
       "      <td>-0.132986</td>\n",
       "      <td>-0.143900</td>\n",
       "      <td>-0.152612</td>\n",
       "      <td>-0.093533</td>\n",
       "      <td>-0.164809</td>\n",
       "      <td>-0.124286</td>\n",
       "      <td>-0.153650</td>\n",
       "      <td>-0.178778</td>\n",
       "      <td>-0.147696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_000010565</th>\n",
       "      <td>GB_GCA_000010565.1</td>\n",
       "      <td>-0.187030</td>\n",
       "      <td>-0.251788</td>\n",
       "      <td>-0.224343</td>\n",
       "      <td>-0.079801</td>\n",
       "      <td>-0.200000</td>\n",
       "      <td>-0.177482</td>\n",
       "      <td>-0.149577</td>\n",
       "      <td>-0.113914</td>\n",
       "      <td>-0.096177</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.101351</td>\n",
       "      <td>-0.181850</td>\n",
       "      <td>-0.177694</td>\n",
       "      <td>-0.190099</td>\n",
       "      <td>-0.100300</td>\n",
       "      <td>-0.222171</td>\n",
       "      <td>-0.170978</td>\n",
       "      <td>-0.203634</td>\n",
       "      <td>-0.148263</td>\n",
       "      <td>-0.179736</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 122 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           accession  PF00380.20  PF00410.20  PF00466.21  \\\n",
       "GB_GCA_000007325  GB_GCA_000007325.1   -0.148160   -0.224515   -0.271998   \n",
       "GB_GCA_000008885  GB_GCA_000008885.1   -0.225820   -0.247991   -0.235449   \n",
       "GB_GCA_000009845  GB_GCA_000009845.1   -0.210141   -0.304301   -0.323290   \n",
       "GB_GCA_000010085  GB_GCA_000010085.1   -0.097661   -0.237436   -0.161992   \n",
       "GB_GCA_000010565  GB_GCA_000010565.1   -0.187030   -0.251788   -0.224343   \n",
       "\n",
       "                  PF01025.20  PF02576.18  PF03726.15  TIGR00006  TIGR00019  \\\n",
       "GB_GCA_000007325   -0.122832   -0.222500   -0.195491  -0.189573  -0.152311   \n",
       "GB_GCA_000008885   -0.223839         NaN   -0.212832  -0.275949  -0.212414   \n",
       "GB_GCA_000009845   -0.130158         NaN         NaN  -0.202842  -0.193418   \n",
       "GB_GCA_000010085   -0.060054   -0.167924   -0.131780  -0.111915  -0.063734   \n",
       "GB_GCA_000010565   -0.079801   -0.200000   -0.177482  -0.149577  -0.113914   \n",
       "\n",
       "                  TIGR00020  ...  TIGR02729  TIGR03263  TIGR03594  TIGR03625  \\\n",
       "GB_GCA_000007325  -0.116115  ...  -0.215168  -0.138229  -0.214850  -0.219437   \n",
       "GB_GCA_000008885  -0.160727  ...  -0.219518  -0.278769  -0.307224  -0.212944   \n",
       "GB_GCA_000009845  -0.160336  ...  -0.205122  -0.180596  -0.214562  -0.172616   \n",
       "GB_GCA_000010085  -0.064409  ...  -0.093497  -0.132986  -0.143900  -0.152612   \n",
       "GB_GCA_000010565  -0.096177  ...  -0.101351  -0.181850  -0.177694  -0.190099   \n",
       "\n",
       "                  TIGR03632  TIGR03654  TIGR03723  TIGR03725  TIGR03953  \\\n",
       "GB_GCA_000007325  -0.161122  -0.178828  -0.207375  -0.232777  -0.213205   \n",
       "GB_GCA_000008885  -0.103641  -0.312473  -0.254505        NaN  -0.265804   \n",
       "GB_GCA_000009845  -0.165642  -0.229449  -0.217595  -0.288547  -0.173903   \n",
       "GB_GCA_000010085  -0.093533  -0.164809  -0.124286  -0.153650  -0.178778   \n",
       "GB_GCA_000010565  -0.100300  -0.222171  -0.170978  -0.203634  -0.148263   \n",
       "\n",
       "                  genome_avg_NOSC  \n",
       "GB_GCA_000007325        -0.223900  \n",
       "GB_GCA_000008885        -0.258189  \n",
       "GB_GCA_000009845        -0.225228  \n",
       "GB_GCA_000010085        -0.147696  \n",
       "GB_GCA_000010565        -0.179736  \n",
       "\n",
       "[5 rows x 122 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge the genome average values into the bac120 nosc dataframe\n",
    "bac120_nosc_df = bac120_nosc_df.merge(reps_genome_nosc_df, left_index=True, right_index=True)\n",
    "bac120_nosc_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/h9/fvnb4xnn63qg71jx6ncvg62w0000gn/T/ipykernel_50445/3351634197.py:2: DtypeWarning: Columns (61,65,74,82,83,85) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  gtdb_genome_metadata = pd.read_csv('../data/gtdb/r207/bac120_metadata_r207.tar.gz',\n"
     ]
    }
   ],
   "source": [
    "# Load metadata about all the genomes\n",
    "gtdb_genome_metadata = pd.read_csv('../data/gtdb/r207/bac120_metadata_r207.tar.gz',\n",
    "                                   compression='gzip', sep='\\t')\n",
    "\n",
    "# Extract the phylogeny from the metadata\n",
    "gtdb_phylo = gtdb_genome_metadata.gtdb_taxonomy.str.split(';', expand=True)\n",
    "gtdb_phylo.columns = ['gtdb_domain', 'gtdb_phylum', 'gtdb_class', 'gtdb_order',\n",
    "                      'gtdb_family', 'gtdb_genus', 'gtdb_species']\n",
    "# Strip the underscores from the names\n",
    "gtdb_phylo = gtdb_phylo.applymap(lambda x: x.split('__')[1] if type(x) == str else x)\n",
    "gtdb_genome_metadata = pd.concat([gtdb_genome_metadata, gtdb_phylo], axis=1)\n",
    "\n",
    "# Rename the first column to \"genome_accession\" -- for some reason defaults to the filename\n",
    "# Also drop the rows with no accession number\n",
    "gtdb_genome_metadata = gtdb_genome_metadata.rename(\n",
    "    columns={gtdb_genome_metadata.columns[0]: 'genome_accession'}).dropna(\n",
    "        axis=0, subset=['genome_accession']).set_index('genome_accession')\n",
    "\n",
    "# For some reason the accession numbers have \".1\" appended to them here \n",
    "# but not in the files defining the bac120 marker genes\n",
    "trimmed_ids = [x.split('.')[0] for x in gtdb_genome_metadata.index.tolist()]\n",
    "gtdb_genome_metadata['trimmed_accession'] = trimmed_ids\n",
    "gtdb_genome_metadata['ncbi_taxid_int'] = gtdb_genome_metadata.ncbi_taxid.astype(int)\n",
    "gtdb_genome_metadata = gtdb_genome_metadata.reset_index().set_index('trimmed_accession')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>J</th>\n",
       "      <td>70</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <td>18</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>6</td>\n",
       "      <td>Transcription</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U</th>\n",
       "      <td>6</td>\n",
       "      <td>Intracellular trafficking, secretion, and vesi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>5</td>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O</th>\n",
       "      <td>5</td>\n",
       "      <td>Posttranslational modification, protein turnov...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>3</td>\n",
       "      <td>Coenzyme transport and metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>3</td>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>2</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>1</td>\n",
       "      <td>Cell cycle control, cell division, chromosome ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>1</td>\n",
       "      <td>Signal transduction mechanisms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     count                                        description\n",
       "COG                                                          \n",
       "J       70    Translation, ribosomal structure and biogenesis\n",
       "L       18              Replication, recombination and repair\n",
       "K        6                                      Transcription\n",
       "U        6  Intracellular trafficking, secretion, and vesi...\n",
       "F        5                Nucleotide transport and metabolism\n",
       "O        5  Posttranslational modification, protein turnov...\n",
       "H        3                  Coenzyme transport and metabolism\n",
       "M        3             Cell wall/membrane/envelope biogenesis\n",
       "C        2                   Energy production and conversion\n",
       "D        1  Cell cycle control, cell division, chromosome ...\n",
       "T        1                     Signal transduction mechanisms"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count up the categories among the bac120\n",
    "cat_counts = bac120_meta_df.COG.value_counts()\n",
    "\n",
    "# Add descriptive annotations\n",
    "cog_df = pd.read_csv('../data/COG-fun-20.csv', index_col=0)\n",
    "descs = cog_df.loc[cat_counts.index].description\n",
    "cat_counts = pd.DataFrame(\n",
    "    dict(count=cat_counts, description=descs))\n",
    "cat_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_cols = bac120_nosc_df.columns[1:-1]\n",
    "mat = bac120_nosc_df[pairwise_cols].values\n",
    "\n",
    "# Need to flatten the matrix to permute rows and cols\n",
    "# correlations will not change if we only shuffle one axis\n",
    "permuted = np.random.permutation(mat.flatten()).reshape(mat.shape)\n",
    "permuted_df = pd.DataFrame(permuted, columns=pairwise_cols)\n",
    "pingouin_corr = pg.pairwise_corr(permuted_df, method='pearson')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot: xlabel='r', ylabel='Count'>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEGCAYAAACKB4k+AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy89olMNAAAACXBIWXMAAAsTAAALEwEAmpwYAAASSUlEQVR4nO3df5BdZX3H8feXVNAZLZJmwUCgiW3oCP5RbaRWS4caO9L0R6IVGqZoOg1NoVhKbS2h/KEzTmZoax2FUZlIGaOoMbW2RItlIIqOMwoGtZFNQhLkhyk7CUEbte1EwG//uCdPL9l7N3d377nn3uz7NbOz9z73nN0Ph5AP58dzTmQmkiQBnNR0AEnS8LAUJEmFpSBJKiwFSVJhKUiSip9qOsBsLFiwIBcvXtx0DEkaKQ888MChzBzr9NlIl8LixYvZvn170zEkaaRExGPdPvPwkSSpsBQkSYWlIEkqLAVJUmEpSJIKS0GSVFgKkqTCUpAkFZaCJKkY6RnN0jC7/IormTh0eNL4wgWncvuttzSQSDo+S0GqycShw4ytuGby+J03NZBG6o2HjyRJhaUgSSosBUlSYSlIkgpLQZJUWAqSpMJSkCQVloIkqXDymjRgu3aOs3zVZZPGnemsYWApSLPU7XYWu/fsZWzF5OWfzpOc6ayhZSlIs9TtdhY7xq9qII00O55TkCQVloIkqbAUJEmFpSBJKiwFSVJhKUiSCktBklRYCpKkwslrUo+mO3NZGkWWgtQjZy5rLrAUpDbd9gbAPQLNDZaC1Kbb3gC4R6C5wRPNkqTCUpAkFbWXQkTMi4hvRsTnqvfzI+LuiNhbfT+tbdnrI2JfRDwUEW+oO5sk6bkGsafw58CutvfrgW2ZuRTYVr0nIs4DVgPnAxcDH4yIeQPIJ0mq1FoKEbEI+C3g1rbhlcCm6vUmYFXb+ObMPJKZjwD7gAvqzCdJeq66rz56H/DXwIvaxs7IzAmAzJyIiNOr8bOAr7Utt78ak+YEn92sYVBbKUTEbwMHM/OBiLiol1U6jGWHn7sOWAdwzjnnzCaiNFR8drOGQZ2Hj14L/G5EPApsBl4XEbcDByJiIUD1/WC1/H7g7Lb1FwFPHPtDM3NjZi7LzGVjY2M1xpekuae2UsjM6zNzUWYupnUC+QuZeTmwFVhTLbYGuKN6vRVYHRGnRMQSYClwf135JEmTNTGj+UZgS0SsBR4HLgHIzPGI2ALsBJ4Brs7MZxvIJ0lz1kBKITPvBe6tXj8FLO+y3AZgwyAySZIm895H0pDzqiQNkqUgDTmvStIgee8jSVJhKUiSCktBklRYCpKkwlKQJBWWgiSpsBQkSYWlIEkqLAVJUmEpSJIKS0GSVFgKkqTCUpAkFZaCJKmwFCRJhaUgSSosBUlSYSlIkgofxymdYC6/4komDh2eNO4zndULS0EaUbt2jrN81WWTxnfv2cuF1948adxnOqsXloI0op7Okxhbcc2k8R3jVzWQRicKzylIkgpLQZJUePhIc1K3k7G79+xlbEUDgaQhYSloTpo4dNjj8VIHHj6SJBWWgiSpsBQkSYWlIEkqLAVJUmEpSJIKS0GSVDhPQSc0J6lJ01NbKUTE84EvA6dUv+fTmfnOiJgPfApYDDwKXJqZ36/WuR5YCzwLXJOZd9WVT3ODk9Sk6alzT+EI8LrM/FFEPA/4SkR8HngTsC0zb4yI9cB64LqIOA9YDZwPnAncExHnZuazNWbUCcI9Aqk/aiuFzEzgR9Xb51VfCawELqrGNwH3AtdV45sz8wjwSETsAy4AvlpXRp043COQ+qPWE80RMS8ivgUcBO7OzPuAMzJzAqD6fnq1+FnAd9tW31+NSZIGpNZSyMxnM/MXgUXABRHx8ikWj04/YtJCEesiYntEbH/yySf7lFSSBAO6JDUz/4vWYaKLgQMRsRCg+n6wWmw/cHbbaouAJzr8rI2ZuSwzl42NjdUZW5LmnNpKISLGIuLF1esXAK8HdgNbgTXVYmuAO6rXW4HVEXFKRCwBlgL315VPkjRZnVcfLQQ2RcQ8WuWzJTM/FxFfBbZExFrgceASgMwcj4gtwE7gGeBqrzySpMGq8+qjHcArOow/BSzvss4GYENdmSRJU/M2F5KkwlKQJBWWgiSpsBQkSYWlIEkqLAVJUtFTKUTEa3sZkySNtl73FG7ucUySNMKmnLwWEb8CvAYYi4i3t33008C8OoNJkgbveDOaTwZeWC33orbxHwBvriuUJKkZU5ZCZn4J+FJEfCQzHxtQJklSQ3q999EpEbGR1nOVyzqZ+bo6Qknqv107x1m+6rJJ4wsXnMrtt97SQCINo15L4Z+AW4BbAe9cKo2gp/Okjo8snbjzpgbSaFj1WgrPZOaHak0iSWpcr5ekfjYi/jQiFkbE/KNftSaTJA1cr3sKR5+U9o62sQRe2t84kqQm9VQKmbmk7iCSpOb1VAoR8dZO45n50f7GkSQ1qdfDR69qe/18Wo/T/AZgKUgjzktV1a7Xw0d/1v4+Ik4FPlZLIkkD5aWqajfTW2f/D7C0n0EkSc3r9ZzCZ2ldbQStG+G9DNhSVyhJUjN6PafwnrbXzwCPZeb+GvJIkhrU0+Gj6sZ4u2ndKfU04Md1hpIkNaPXJ69dCtwPXAJcCtwXEd46W5JOML0eProBeFVmHgSIiDHgHuDTdQWTJA1er1cfnXS0ECpPTWNdSdKI6HVP4d8j4i7gk9X73wfurCeSJKkpx3tG888DZ2TmOyLiTcCvAgF8Ffj4APJJkgboeIeA3gf8ECAzP5OZb8/Mv6C1l/C+eqNJkgbteKWwODN3HDuYmdtpPZpTknQCOd45hedP8dkL+hlE6sXlV1zJxKHDk8Z379nL2IoGAkknmOOVwtcj4o8z88PtgxGxFnigvlhSZxOHDne8eduO8asaSCOdeI5XCtcC/xIRf8D/l8Ay4GTgjTXmkiQ1YMpSyMwDwGsi4teBl1fD/5aZX6g9mSRp4Hp9nsIXgS/WnEWS1LBeJ69NW0ScTevJbC8BfgJszMz3R8R84FO0rl56FLg0M79frXM9sBZ4FrgmM++qK5+GmyeUpWbUVgq0brH9l5n5jYh4EfBARNwN/CGwLTNvjIj1wHrguog4D1gNnA+cCdwTEedm5rM1ZtSQ8oSy1IzaSiEzJ4CJ6vUPI2IXcBawErioWmwTcC9wXTW+OTOPAI9ExD7gAlqzp3WCco9AGi517ikUEbEYeAVwH63bZhwti4mIOL1a7Czga22r7a/Gjv1Z64B1AOecc06NqTUI7hFIw6X2O51GxAuBfwauzcwfTLVoh7GcNJC5MTOXZeaysbGxfsWUJFFzKUTE82gVwscz8zPV8IGIWFh9vhA4ekvu/cDZbasvAp6oM58k6bnqvPoogH8EdmXme9s+2gqsAW6svt/RNv6JiHgvrRPNS2k97U1SA3btHGf5qss6frZwwancfustA06kQajznMJrgbcA346Ib1Vjf0OrDLZUt8p4nNYjPsnM8YjYAuykdeXS1V55JDXn6Typ4/kegIk7bxpwGg1KnVcffYXO5wkAlndZZwOwoa5MkqSp+UhNSVJhKUiSCktBklRYCpKkwlKQJBWWgiSpsBQkSYWlIEkqLAVJUmEpSJIKS0GSVFgKkqTCUpAkFZaCJKkYyDOaJZ1Yuj2Ax4fvjD5LQdK0dXsAjw/fGX0ePpIkFZaCJKmwFCRJhaUgSSosBUlSYSlIkgpLQZJUOE9BUt84qW30WQqS+sZJbaPPw0eSpMJSkCQVloIkqbAUJEmFpSBJKrz6SANx+RVXMnHo8KTx3Xv2MraigUCSOrIUNBAThw53vFRxx/hVDaSR1I2HjyRJhaUgSSosBUlSUVspRMRtEXEwIh5sG5sfEXdHxN7q+2ltn10fEfsi4qGIeENduSRJ3dW5p/AR4OJjxtYD2zJzKbCtek9EnAesBs6v1vlgRMyrMZskqYPaSiEzvwx875jhlcCm6vUmYFXb+ObMPJKZjwD7gAvqyiZJ6mzQ5xTOyMwJgOr76dX4WcB325bbX41NEhHrImJ7RGx/8sknaw0rSXPNsMxTiA5j2WnBzNwIbARYtmxZx2XUHCepSaNt0KVwICIWZuZERCwEDlbj+4Gz25ZbBDwx4Gyahqn+8r/w2psnjTtJbW7z4TujY9ClsBVYA9xYfb+jbfwTEfFe4ExgKXD/gLNpGpyhrOnw4Tujo7ZSiIhPAhcBCyJiP/BOWmWwJSLWAo8DlwBk5nhEbAF2As8AV2fms3VlkyR1VlspZObkfcWW5V2W3wBsqCuPJOn4nNEsSSosBUlSYSlIkophmacgaQ7yUtXhYylIaoyXqg4fDx9JkgpLQZJUWAqSpMJSkCQVloIkqfDqI03JW2FLc4uloCl5N1RpbvHwkSSpcE9B0tBxpnNzLAVJQ8eZzs2xFCSNDPcg6mcpSBoZ7kHUzxPNkqTCUpAkFR4+EuAkNUktloIAJ6lJarEU5pBuewPgHoGkFkthDum2NwDuEUhqsRQkjTznL/SPpSBp5Dl/oX+8JFWSVFgKkqTCw0cnIOccSJopS+EE5JwDSTNlKYww9wikqXlV0vRZCiPMPQJpal6VNH2Wwghwj0DqL/cgurMURoB7BFJ/uQfRnZekSpIK9xSGiIeJJDVt6EohIi4G3g/MA27NzBsbjjQwHiaS1LShKoWImAd8APgNYD/w9YjYmpk7m002M93+z9+TWdJw6nYC+tGH97D4587teXwm6wzL3wtDVQrABcC+zPwOQERsBlYCtZRCt7+0p/svs9v47j17ufDamyeNf+E9f9LxD56HiaRmdTsBvePvr5rW+EzW6fb3wqBLJDKz7z90piLizcDFmXlF9f4twC9n5tvallkHrKve/gLw0MCD9m4BcKjpED0apawwWnlHKSuMVt5RygrDk/dnM3Os0wfDtqcQHcae01qZuRHYOJg4sxMR2zNzWdM5ejFKWWG08o5SVhitvKOUFUYj77BdkrofOLvt/SLgiYaySNKcM2yl8HVgaUQsiYiTgdXA1oYzSdKcMVSHjzLzmYh4G3AXrUtSb8vM8YZjzcZIHOaqjFJWGK28o5QVRivvKGWFEcg7VCeaJUnNGrbDR5KkBlkKkqTCUpimiJgfEXdHxN7q+2ldlrs4Ih6KiH0Rsb5t/JKIGI+In0TEsrbxxRHxvxHxreqrL7NS6spbfXZ9tfxDEfGGIcjacf1+bttuv7vt84iIm6rPd0TEK2eaux9qyvuuiPjPtu3ZtymXs8x7W0QcjIgHj1mnlu1bU9batm3PMtOvaXwBfwesr16vB/62wzLzgIeBlwInA/8BnFd99jJak+7uBZa1rbMYeHCE8p5XLXcKsKRaf17DWTuu369tO9XvbltmBfB5WnNuXg3cN9PcQ5z3XcBf1fBndcZ5q89+DXjlsf+u69i+NWatZdtO58s9helbCWyqXm8CVnVYptyuIzN/DBy9XQeZuSszBzkLu668K4HNmXkkMx8B9lU/p7GsPa4/G1P97qNWAh/Nlq8BL46IhQ3lritvXWaTl8z8MvC9Dj+3ju1bV9bGWQrTd0ZmTgBU30/vsMxZwHfb3u+vxo5nSUR8MyK+FBEXzj4qUF/emf4zTmW2Wadavx/btpd/5m7LzDT3bNSVF+Bt1SGR2/p4uGs2eadSx/atKyvUs217NlTzFIZFRNwDvKTDRzf0+iM6jB3v2t8J4JzMfCoifgn414g4PzN/cNxf1kzemawzctt2Br+72zIz2l6zVFfeDwHvrt6/G/gH4I9mmLGXLNNdZhDqylrXtu2ZpdBBZr6+22cRcSAiFmbmRLUreLDDYtO+XUdmHgGOVK8fiIiHgXOB7cOYd4br1J214/qz2bbT+N3HW+bk6ebug1ryZuaBo4MR8WHgc0OQdyp1bN9asta4bXvm4aPp2wqsqV6vAe7osMy0b9cREWPRep4EEfFSYCnwnWHNW32+OiJOiYglVd77G87acf0+bttettNW4K3VlSevBg5XhyymnbsPasl79Lh45Y3Ag/THbPJOpY7tW0vWGrdt75o8yz2KX8DPANuAvdX3+dX4mcCdbcutAPbQukLhhrbxN9L6P4gjwAHgrmr894BxWlcxfAP4nWHOW312Q7X8Q8BvDkHWbuv3bdt2+t3AlcCV1eug9aCoh4Fv89wrtqaVu0///uvI+7Fq2R20/uJbOCR5P0nrUOHT1Z/ZtXVu35qy1rZte/3yNheSpMLDR5KkwlKQJBWWgiSpsBQkSYWlIEkqLAVJUmEpSDWrJi/535pGgn9QpRpE6xkOuyLig7QmzJ19vHWkYeDkNakGEbGY1q00XpOt2yZLI8E9Bak+j1kIGjWWglSf/246gDRdloIkqbAUJEmFJ5olSYV7CpKkwlKQJBWWgiSpsBQkSYWlIEkqLAVJUmEpSJKK/wNPhEohe3gaBAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Permutation totally erases the correlation structure\n",
    "sns.histplot(pingouin_corr.r, bins=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# last column is genome_avg_NOSC -- \n",
    "# calculate the pairwise correlations of columns except GC\n",
    "pairwise_cols = bac120_nosc_df.columns[:-1]\n",
    "nosc_corr = pg.pairwise_corr(\n",
    "    bac120_nosc_df, columns=pairwise_cols,\n",
    "    padjust='fdr_bh')\n",
    "\n",
    "# Now calculate the partial correlations, controlling for genome_avg_NOSC\n",
    "nosc_corr_controlled = pg.pairwise_corr(\n",
    "    bac120_nosc_df, columns=pairwise_cols, covar='genome_avg_NOSC',\n",
    "    padjust='fdr_bh')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = lambda row: ','.join(sorted(row))\n",
    "def _save_corr_df(corr_df, fname):\n",
    "    # Add descriptions of the two columns\n",
    "    corr_df['X_desc'] = bac120_meta_df.loc[corr_df.X].Description.values\n",
    "    corr_df['Y_desc'] = bac120_meta_df.loc[corr_df.Y].Description.values\n",
    "\n",
    "    # Add COG categories of the two cols\n",
    "    corr_df['X_COG'] = bac120_meta_df.loc[corr_df.X].COG.values\n",
    "    corr_df['Y_COG'] = bac120_meta_df.loc[corr_df.Y].COG.values\n",
    "\n",
    "    # Order-independent COG-pair by sorting the high-level COG-categories\n",
    "    \n",
    "    corr_df['COG_pair'] = list(map(f, corr_df['X_COG,Y_COG'.split(',')].values))\n",
    "    \n",
    "    # Save\n",
    "    corr_df.to_csv(fname, index=False)\n",
    "\n",
    "    \n",
    "my_fname = '../data/gtdb/r207/bac120_nosc_corr.csv'\n",
    "_save_corr_df(nosc_corr, my_fname)\n",
    "\n",
    "my_fname = '../data/gtdb/r207/bac120_nosc_corr_controlled_for_genome_nosc.csv'\n",
    "_save_corr_df(nosc_corr_controlled, my_fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X</th>\n",
       "      <th>Y</th>\n",
       "      <th>method</th>\n",
       "      <th>alternative</th>\n",
       "      <th>n</th>\n",
       "      <th>r</th>\n",
       "      <th>CI95%</th>\n",
       "      <th>p-unc</th>\n",
       "      <th>p-corr</th>\n",
       "      <th>p-adjust</th>\n",
       "      <th>BF10</th>\n",
       "      <th>power</th>\n",
       "      <th>X_desc</th>\n",
       "      <th>Y_desc</th>\n",
       "      <th>X_COG</th>\n",
       "      <th>Y_COG</th>\n",
       "      <th>COG_pair</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3487</th>\n",
       "      <td>TIGR00344</td>\n",
       "      <td>TIGR00392</td>\n",
       "      <td>pearson</td>\n",
       "      <td>two-sided</td>\n",
       "      <td>51309</td>\n",
       "      <td>0.817759</td>\n",
       "      <td>[0.81, 0.82]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>fdr_bh</td>\n",
       "      <td>inf</td>\n",
       "      <td>1.0</td>\n",
       "      <td>alanine--tRNA ligase</td>\n",
       "      <td>isoleucine--tRNA ligase</td>\n",
       "      <td>J</td>\n",
       "      <td>J</td>\n",
       "      <td>J,J</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3747</th>\n",
       "      <td>TIGR00392</td>\n",
       "      <td>TIGR00456</td>\n",
       "      <td>pearson</td>\n",
       "      <td>two-sided</td>\n",
       "      <td>50441</td>\n",
       "      <td>0.813436</td>\n",
       "      <td>[0.81, 0.82]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>fdr_bh</td>\n",
       "      <td>inf</td>\n",
       "      <td>1.0</td>\n",
       "      <td>isoleucine--tRNA ligase</td>\n",
       "      <td>arginine--tRNA ligase</td>\n",
       "      <td>J</td>\n",
       "      <td>J</td>\n",
       "      <td>J,J</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3229</th>\n",
       "      <td>TIGR00194</td>\n",
       "      <td>TIGR00392</td>\n",
       "      <td>pearson</td>\n",
       "      <td>two-sided</td>\n",
       "      <td>50982</td>\n",
       "      <td>0.811021</td>\n",
       "      <td>[0.81, 0.81]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>fdr_bh</td>\n",
       "      <td>inf</td>\n",
       "      <td>1.0</td>\n",
       "      <td>excinuclease ABC subunit C</td>\n",
       "      <td>isoleucine--tRNA ligase</td>\n",
       "      <td>L</td>\n",
       "      <td>J</td>\n",
       "      <td>J,L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3252</th>\n",
       "      <td>TIGR00194</td>\n",
       "      <td>TIGR00634</td>\n",
       "      <td>pearson</td>\n",
       "      <td>two-sided</td>\n",
       "      <td>47523</td>\n",
       "      <td>0.810044</td>\n",
       "      <td>[0.81, 0.81]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>fdr_bh</td>\n",
       "      <td>inf</td>\n",
       "      <td>1.0</td>\n",
       "      <td>excinuclease ABC subunit C</td>\n",
       "      <td>DNA repair protein RecN</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>L,L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5518</th>\n",
       "      <td>TIGR00643</td>\n",
       "      <td>TIGR01391</td>\n",
       "      <td>pearson</td>\n",
       "      <td>two-sided</td>\n",
       "      <td>48403</td>\n",
       "      <td>0.800016</td>\n",
       "      <td>[0.8, 0.8]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>fdr_bh</td>\n",
       "      <td>inf</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ATP-dependent DNA helicase RecG</td>\n",
       "      <td>DNA primase</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>L,L</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              X          Y   method alternative      n         r  \\\n",
       "3487  TIGR00344  TIGR00392  pearson   two-sided  51309  0.817759   \n",
       "3747  TIGR00392  TIGR00456  pearson   two-sided  50441  0.813436   \n",
       "3229  TIGR00194  TIGR00392  pearson   two-sided  50982  0.811021   \n",
       "3252  TIGR00194  TIGR00634  pearson   two-sided  47523  0.810044   \n",
       "5518  TIGR00643  TIGR01391  pearson   two-sided  48403  0.800016   \n",
       "\n",
       "             CI95%  p-unc  p-corr p-adjust BF10  power  \\\n",
       "3487  [0.81, 0.82]    0.0     0.0   fdr_bh  inf    1.0   \n",
       "3747  [0.81, 0.82]    0.0     0.0   fdr_bh  inf    1.0   \n",
       "3229  [0.81, 0.81]    0.0     0.0   fdr_bh  inf    1.0   \n",
       "3252  [0.81, 0.81]    0.0     0.0   fdr_bh  inf    1.0   \n",
       "5518    [0.8, 0.8]    0.0     0.0   fdr_bh  inf    1.0   \n",
       "\n",
       "                               X_desc                   Y_desc X_COG Y_COG  \\\n",
       "3487             alanine--tRNA ligase  isoleucine--tRNA ligase     J     J   \n",
       "3747          isoleucine--tRNA ligase    arginine--tRNA ligase     J     J   \n",
       "3229       excinuclease ABC subunit C  isoleucine--tRNA ligase     L     J   \n",
       "3252       excinuclease ABC subunit C  DNA repair protein RecN     L     L   \n",
       "5518  ATP-dependent DNA helicase RecG              DNA primase     L     L   \n",
       "\n",
       "     COG_pair  \n",
       "3487      J,J  \n",
       "3747      J,J  \n",
       "3229      J,L  \n",
       "3252      L,L  \n",
       "5518      L,L  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nosc_corr.sort_values('r', ascending=False).head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda.base",
   "language": "python",
   "name": "conda.base"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
